---
title: "Prepare files for week 2 and 3"
author: "Christa Toenhake"
date: "13/11/2020"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE, echo = T, eval = F}
knitr::opts_chunk$set(, echo = TRUE)
```

### Load libraries 

```{r libraries, message=FALSE, warning=FALSE, echo = T, eval = F}
#library(tidyverse)
library(GenomicRanges)
library(rtracklayer)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
library(stringr)
library(plyranges)
#BiocManager::install("biomaRt")
library(biomaRt)
#BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)
#BiocManager::install("reader") # to remove extension from bw files names --> only installed on my own mac
#library(reader)
```


### chr19 peaks in GRanges 
```{r prepare monocyt data, echo = T, eval = F}
# H3k4me1 
mono_h3k4me1 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547981.H3K4me1.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K4me3  
mono_h3k4me3 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547984.H3K4me3.bwa.GRCh38.20150527.chr19.bed", format = "narrowPeak")
# H3K9me3 
mono_h3k9me3 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547982.H3K9me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K27ac 
mono_h3k27ac <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547980.H3K27ac.bwa.GRCh38.20150527.chr19.bed", format = "narrowPeak")
# H3K27me3   
mono_h3k27me3 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547983.H3K27me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K36me3
mono_h3k36me3 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547979.H3K36me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# create list
mono_list  <- GRangesList(mono_h3k4me1, mono_h3k4me3, mono_h3k9me3, mono_h3k27ac, mono_h3k27me3, mono_h3k36me3)
# add names to each element in the list
names(mono_list) <- c("h3k4me1", "h3k4me3", "h3k9me3", "h3k27ac", "h3k27me3", "h3k36me3")
```

```{r save prepare monocyt data, echo = T, eval = F}
save(mono_h3k4me1, mono_h3k4me3, mono_h3k9me3, mono_h3k27ac, mono_h3k27me3, mono_h3k36me3, mono_list, file =  "data/prepared_rds/blueprint_monocyte_chr19_granges.RData")
```


### chr19 BLUEPRINT monocytes transcript quantification  

- add entrez gene id using org.Hs.eg.db package   
- restrict to genes on chromosome 19  

```{r quantification, echo = T, eval = F}
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)

all_results <- read.table("data/blueprint/rna_quantification/C000S5B1.gene_quantification.rsem_grape2_crg.GRCh38.20150622.results", header =T)

# remove version from ensemble gene id 
all_results$gene_id <- sapply(all_results$gene_id,  sub, pattern = "\\.\\d+$", replacement = "")

# map entrez id to ensemble id
all_results$entrezgene_id = mapIds(org.Hs.eg.db,
                    keys=all_results$gene_id, 
                    column="ENTREZID",
                    keytype="ENSEMBL",
                    multiVals="first")

# vector of entrez ids for genes on chr19 
genes_chr19 <- genes(txdb)
geneids_genes_chr19 <- mcols(genes_chr19)[,1]

# subset all_results for genes on chr19
quantification_chr19 <- all_results[all_results$entrezgene_id %in% geneids_genes_chr19,c("entrezgene_id", "TPM", "FPKM")]
```

```{r save quantification, echo = T, eval = F}
# write quantification_chr19 to rds
saveRDS(quantification_chr19, file = "data/prepared_rds/blueprint_c000s5_gene_quantification_chr19.rds")

```

### chr19 Complex Heatmap object week 2  
Used for exercises in fg2 (2021 version)  
This function is also run by students but resulting object by several exercises and as the function takes a while, I decided to prepare the object in advance.  
```{r, echo = T, eval = F}
# prepare objects for upset plot in week 2
monocytes_combinationmatrix <- make_comb_mat(mono_list) # takes a while! 

saveRDS(monocytes_combinationmatrix, file =  "data/prepared_rds/blueprint_monocyte_chr19_complexheatmap_combmatrix.rds")

```

## chr19 - get smaller bigwig by downsampling genes and save associated tss and FPKM files    
Problem: chip-seq bigwig files of 1 chromosome are too large  
Aim: for 5MB per tutorial on data.  
Possible solutions:  

1. Select chr19, only 20kb +/- TSS sites \-\-\> files remain too large   
2. Work with chr21 \-\-\> files are small enough but gene expression is not representative for genome wide  
3. Random select subsample of genes from chr19 based on quantification quartiles and used these to select regions in bigwig file.    

```{r step 1 chr19 subsample per gene expression quantile, eval = F}
# Important parameter: the chosen proportion of genes
chosen_proportion <- 0.4
```

```{r step 2 chr19 subsample per gene expression quantile, eval = F}
### set sqlevels in txdb to chr19
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
seqlevels(txdb) <- seqlevels0(txdb)
seqlevels(txdb) <- "chr19"

# load chr19 quantification (used in fg2)
rdsfile <- system.file("extdata", "week2", "prepared_rds", "blueprint_c000s5_gene_quantification_chr19.rds", package = "learnr.proto")
quantification_chr19 <- readRDS(rdsfile)
```

```{r step 3 chr19 subsample per gene expression quantile, eval = F}
# define new variable based on quartiles of gene expression
quantification_chr19 <- quantification_chr19 %>% 
  mutate(expression_group = ntile(FPKM, 4))

### set seed for slice_sample
set.seed(13)

# subsample per quartile  
quantification_chr19_sample <- quantification_chr19 %>% 
  group_by(expression_group) %>% 
  slice_sample(prop=chosen_proportion, replace = F)

# check number of genes per quantile and in expression plot
table(quantification_chr19_sample$expression_group)
ggplot(quantification_chr19, aes(x = as.factor(expression_group), y = log10(FPKM+1)))+
  geom_boxplot()+
  geom_jitter()
ggplot(quantification_chr19_sample, aes(x = as.factor(expression_group), y = log10(FPKM+1)))+
  geom_boxplot()+
  geom_jitter()
```

```{r setp 4 subsample bigiwg file for these TSS windows, eval = F}
# subsample bigwig for regions around the TSS of the subsampled genes on chr19  
## set fref files and libraries:
tss_chr19 <- unique(promoters(genes(txdb), upstream = 2000, downstream = 0 )) 
tss_chr19 <- resize(tss_chr19, width = 1, fix = "end")
tss_chr19_sample <- tss_chr19 %>% filter(gene_id %in% quantification_chr19_sample$entrezgene_id )  

## make sure the same geneIDs are present in tss_chr19_sample and quantification_chr19_sample 
tss_genes <- tss_chr19_sample$gene_id
quant_genes <- quantification_chr19_sample$entrezgene_id

which(!(quant_genes %in% tss_genes))
quant_genes[246]
quantification_chr19_sample <- quantification_chr19_sample[-246,]
which(!(quantification_chr19_sample$entrezgene_id %in% tss_genes))

### input dir:
dir_raw_files <- "/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19/"
### output dir:
dir_out_files <- "/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19_tss_sample/"

### function to select regions in bigwig files and write output in outdir
bw_files <- dir(path = dir_raw_files, pattern = "*.bw")
resize_bw <- function(w_in_kb, f, tss){
  print(deparse(substitute(f)))
  w_in_bp <- w_in_kb*1000+4
  tss_window <- resize(tss, width = w_in_bp, fix = "center")
  sum(width(tss_window))
  bw_tss_window <- import(paste(dir_raw_files, f, sep = "/"), which = tss_window, as = "RleList")
  bw_out <- file.path(dir_out_files, 
                      paste0(tools::file_path_sans_ext(f), "_chr19tss_", chosen_proportion, "ofGenes_", w_in_kb, "kb.bw"))
  export.bw(object = bw_tss_window, bw_out)
  rm(tss_window, bw_tss_window, bw_out)
}

## perform function on all files in input directory. 
for(file in bw_files){
resize_bw(10, file, tss_chr19_sample)
}

# make quantification_chr19_sample ready for saving
quantification_chr19_sample$expression_group <- NULL
quantification_chr19_sample <- quantification_chr19_sample %>% dplyr::rename(gene_id = entrezgene_id)
  
# make tss_chr19_sample ready for saving, thus add FPKM to TSS windows such that these intervals can be arranged by FPKM
FPKM_tssorder<- as.vector(NA)
for(i in 1:length(tss_chr19_sample)) {
  FPKM_tssorder[i] <- quantification_chr19_sample[which(quantification_chr19_sample[, "gene_id"] == tss_chr19_sample$gene_id[i]),]$FPKM
  }
tss_chr19_sample$FPKM <- FPKM_tssorder
rm(FPKM_tssorder)
  
## save sampled/sliced quantification and TSSs
save(quantification_chr19_sample, tss_chr19_sample, file = "/home/ctoenhake/learnr.proto/inst/extdata/week3.Rdata")
```

### session Info
```{r session info, echo = T, eval = F}
sessionInfo()

```

