---
title: "Prepare files for week 2 and 3"
author: "Christa Toenhake"
date: "13/11/2020"
output: html_document
---

### Last updated 2021/02/06

```{r setup, include=FALSE, echo = T}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

### Load libraries 

```{r libraries, message=FALSE, warning=FALSE, echo = T, eval = F}
#library(tidyverse)
library(GenomicRanges)
library(rtracklayer)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
library(stringr)
library(plyranges)
#BiocManager::install("biomaRt")
#BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)
#BiocManager::install("reader") # to remove extension from bw files names --> only installed on my own mac
#library(reader)
```

# week 2

### chr19 peaks in GRanges   
the raw data are removed from the tutorial package and can be found on blueprint dataportal and on cn45:   

- bed files were downloaded from, eg.H3K4me3: http://dcc.blueprint-epigenome.eu/#/experiments/ERX547984 ; file "Enriched regions, BED format'.  
- select chr19 with awk:  `cat $file | awk '{OFS = "\t"} {if ($1 == "chr19") print $0}'  > $(basename $file .bed).chr19.bed`   
- complete and chr19 limited .bed files are stored on cn45: `/ceph/rimlsfnwi/web_share/mbdata/ctoenhake/edu/fg1_r_data/blueprint/chip_bed`  

```{r prepare monocyt data, echo = T, eval = F}
# H3k4me1 
monocytes_h3k4me1 <- rtracklayer::import("/home/ctoenhake/learnr.proto/inst/extdata/week2/blueprint/chip_bed/C000S5H2.ERX547981.H3K4me1.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K4me3  
monocytes_h3k4me3 <- rtracklayer::import("/home/ctoenhake/learnr.proto/inst/extdata/week2/blueprint/chip_bed/C000S5H2.ERX547984.H3K4me3.bwa.GRCh38.20150527.chr19.bed", format = "narrowPeak")
# H3K9me3 
monocytes_h3k9me3 <- rtracklayer::import("/home/ctoenhake/learnr.proto/inst/extdata/week2/blueprint/chip_bed/C000S5H2.ERX547982.H3K9me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K27ac 
monocytes_h3k27ac <- rtracklayer::import("/home/ctoenhake/learnr.proto/inst/extdata/week2/blueprint/chip_bed/C000S5H2.ERX547980.H3K27ac.bwa.GRCh38.20150527.chr19.bed", format = "narrowPeak")
# H3K27me3   
monocytes_h3k27me3 <- rtracklayer::import("/home/ctoenhake/learnr.proto/inst/extdata/week2/blueprint/chip_bed/C000S5H2.ERX547983.H3K27me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K36me3
monocytes_h3k36me3 <- rtracklayer::import("/home/ctoenhake/learnr.proto/inst/extdata/week2/blueprint/chip_bed/C000S5H2.ERX547979.H3K36me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# create list
monocytes_list  <- GRangesList(monocytes_h3k4me1, monocytes_h3k4me3, monocytes_h3k9me3, monocytes_h3k27ac, monocytes_h3k27me3, monocytes_h3k36me3)
# add names to each element in the list
names(monocytes_list) <- c("h3k4me1", "h3k4me3", "h3k9me3", "h3k27ac", "h3k27me3", "h3k36me3")
```

```{r save prepare monocyt data, echo = T, eval = F}
save(monocytes_h3k4me1, monocytes_h3k4me3, monocytes_h3k9me3, monocytes_h3k27ac, monocytes_h3k27me3, monocytes_h3k36me3, monocytes_list,  file =  "data/prepared_rds/blueprint_monocyte_chr19_granges.RData")
```


### chr19 BLUEPRINT monocytes transcript quantification  

- add entrez gene id using org.Hs.eg.db package   
- restrict to genes on chromosome 19  

- File "C000S5B1.gene_quantification.rsem_grape2_crg.GRCh38.20150622.results" 
  + was downloaded from:  http://dcc.blueprint-epigenome.eu/#/experiments/ERX157053 ; file "Transcription quantification (Genes)".  
  + is stored on cn45: `/ceph/rimlsfnwi/web_share/mbdata/ctoenhake/edu/fg1_r_data/blueprint/quant/C000S5B1.gene_quantification.rsem_grape2_crg.GRCh38.20150622.results`.   

```{r quantification, echo = T, eval = F}
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)

all_results <- read.table("data/blueprint/rna_quantification/C000S5B1.gene_quantification.rsem_grape2_crg.GRCh38.20150622.results", header =T)

# remove version from ensemble gene id 
all_results$gene_id <- sapply(all_results$gene_id,  sub, pattern = "\\.\\d+$", replacement = "")

# map entrez id to ensemble id
all_results$entrezgene_id = mapIds(org.Hs.eg.db,
                    keys=all_results$gene_id, 
                    column="ENTREZID",
                    keytype="ENSEMBL",
                    multiVals="first")

# vector of entrez ids for genes on chr19 
genes_chr19 <- genes(txdb)
geneids_genes_chr19 <- mcols(genes_chr19)[,1]

# subset all_results for genes on chr19
quantification_chr19 <- all_results[all_results$entrezgene_id %in% geneids_genes_chr19,c("entrezgene_id", "TPM", "FPKM")]
```

```{r save quantification, echo = T, eval = F}
# write quantification_chr19 to rds
saveRDS(quantification_chr19, file = "data/prepared_rds/blueprint_c000s5_gene_quantification_chr19.rds")

```

# week 3

- Situation: need bigWig files for practical.  
- Problem: these files are too large, will take long time before practical is loaded.   
- Target: files to be used should not exceed 5MB (preferably for the whole package but that is too little).  
- Failed solution1: select chromosome 19 only --> but this leaves us with bigwig files of Â±20-30 MB each! Still too large.  
- Failed solution2: work with chr21 \-\-\> files are small enough but gene expression is not representative for genome wide  
- Successfull solution3: Random select 40% of genes from chr19 based on quantification quartiles and used these to select regions in bigwig file.    

### How I obtained the bigwig files used in this script (covering chgr19 only):  

- bigwig files were downloaded from, eg.H3K4me3: http://dcc.blueprint-epigenome.eu/#/experiments/ERX547984 ; file "Normalized signal, BigWig format'.  
- select chr19 with script: `/scratch/ctoenhake/data/fg1_r_data/blueprint/chip_bw/select_chr19.sh`  
- used a bit of a work-around to get these bigwigs on passer:  
    + copy resulting *.chr19.bw to the web accessible directory that I also used for the trackhubs `/ceph/rimlsfnwi/web_share/mbdata/ctoenhake/edu/fg1_r_data/blueprint/chip_bw`  
    + Download file from there to my mac with curl -O [filename] .
    + Upload these files to the learn.proto `/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19`  

After I ran the chunks to get smaller bigwig files, I removed the dir `/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19`, and kept `/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19_tss_sample`   


## chr19 - get smaller bigwig by downsampling genes and save associated tss and FPKM files    

```{r step 1 chr19 subsample per gene expression quantile, eval = F}
# Important parameter: the chosen proportion of genes
chosen_proportion <- 0.4
```

```{r step 2 chr19 subsample per gene expression quantile, eval = F}
### set sqlevels in txdb to chr19
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
seqlevels(txdb) <- seqlevels0(txdb)
seqlevels(txdb) <- "chr19"

# load chr19 quantification (used in fg2)
rdsfile <- system.file("extdata", "week2", "prepared_rds", "blueprint_c000s5_gene_quantification_chr19.rds", package = "learnr.proto")
quantification_chr19 <- readRDS(rdsfile)
```

```{r step 3 chr19 subsample per gene expression quantile, eval = F}
# define new variable based on quartiles of gene expression
quantification_chr19 <- quantification_chr19 %>% 
  mutate(expression_group = ntile(FPKM, 4))

### set seed for slice_sample
set.seed(13)

# subsample per quartile  
quantification_chr19_sample <- quantification_chr19 %>% 
  group_by(expression_group) %>% 
  slice_sample(prop=chosen_proportion, replace = F)

# check number of genes per quantile and in expression plot
table(quantification_chr19_sample$expression_group)
ggplot(quantification_chr19, aes(x = as.factor(expression_group), y = log10(FPKM+1)))+
  geom_boxplot()+
  geom_jitter()
ggplot(quantification_chr19_sample, aes(x = as.factor(expression_group), y = log10(FPKM+1)))+
  geom_boxplot()+
  geom_jitter()
```

```{r setp 4 subsample bigiwg file for these TSS windows, eval = F}
# subsample bigwig for regions around the TSS of the subsampled genes on chr19  
## set fref files and libraries:
tss_chr19 <- unique(promoters(genes(txdb), upstream = 2000, downstream = 0 )) 
tss_chr19 <- resize(tss_chr19, width = 1, fix = "end")
tss_chr19_sample <- tss_chr19 %>% filter(gene_id %in% quantification_chr19_sample$entrezgene_id )  

## make sure the same geneIDs are present in tss_chr19_sample and quantification_chr19_sample 
tss_genes <- tss_chr19_sample$gene_id
quant_genes <- quantification_chr19_sample$entrezgene_id

which(!(quant_genes %in% tss_genes))
quant_genes[246]
quantification_chr19_sample <- quantification_chr19_sample[-246,]
which(!(quantification_chr19_sample$entrezgene_id %in% tss_genes))

### input dir:
dir_raw_files <- "/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19/"
### output dir:
dir_out_files <- "/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19_tss_sample/"

### function to select regions in bigwig files and write output in outdir
bw_files <- dir(path = dir_raw_files, pattern = "*.bw")
resize_bw <- function(w_in_kb, f, tss){
  print(deparse(substitute(f)))
  w_in_bp <- w_in_kb*1000+4
  tss_window <- resize(tss, width = w_in_bp, fix = "center")
  sum(width(tss_window))
  bw_tss_window <- import(paste(dir_raw_files, f, sep = "/"), which = tss_window, as = "RleList")
  bw_out <- file.path(dir_out_files, 
                      paste0(tools::file_path_sans_ext(f), "_chr19tss_", chosen_proportion, "ofGenes_", w_in_kb, "kb.bw"))
  export.bw(object = bw_tss_window, bw_out)
  rm(tss_window, bw_tss_window, bw_out)
}

## perform function on all files in input directory. 
for(file in bw_files){
resize_bw(10, file, tss_chr19_sample)
}
```

# chr19 - save objects in week3.Rdata
```{r objects needed in week3, eval =FALSE, echo=TRUE}  
# make quantification_chr19_sample (generated in chunk above) ready for saving
quantification_chr19_sample$expression_group <- NULL
quantification_chr19_sample <- quantification_chr19_sample %>% dplyr::rename(gene_id = entrezgene_id)
  
# make tss_chr19_sample  (generated in chunk above)  ready for saving, thus add FPKM to TSS windows such that these intervals can be arranged by FPKM
FPKM_tssorder<- as.vector(NA)
for(i in 1:length(tss_chr19_sample)) {
  FPKM_tssorder[i] <- quantification_chr19_sample[which(quantification_chr19_sample[, "gene_id"] == tss_chr19_sample$gene_id[i]),]$FPKM
  }
tss_chr19_sample$FPKM <- FPKM_tssorder
rm(FPKM_tssorder)

# load week2 data and restrict  to granges objects that are used
rdata = system.file("extdata", "week2.Rdata", package = "learnr.proto")
load(rdata)

# load the txdb package which holds transcript-based gene models of hg38 genome  
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
seqlevels(txdb) <- "chr19"

# obtain tss, used to identify nearest TSS for each H3k4me3 peak
# add column with 'name', this column is recognized by the 'distance2NearestFeature' function used in week3
tss_chr19 <- unique(promoters(genes(txdb), upstream=100, downstream=0)) 
tss_chr19  <- resize(tss_chr19, width = 1, fix = "end")
tss_chr19$name <- tss_chr19$gene_id

## save granges objects of peaks and sampled/sliced quantification and TSSs
save(monocytes_h3k4me1, monocytes_h3k4me3, monocytes_h3k9me3, monocytes_h3k27ac, monocytes_h3k27me3, monocytes_h3k36me3,  tss_chr19, quantification_chr19_sample, tss_chr19_sample, file = "/home/ctoenhake/learnr.proto/inst/extdata/week3.Rdata")
```


### session Info
```{r session info, echo = T, eval = T}
sessionInfo()

```

