---
title: "Prepare files for week 2 and 3"
author: "Christa Toenhake"
date: "13/11/2020"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE, echo = T, eval = F}
knitr::opts_chunk$set(, echo = TRUE)
```

### Load libraries 

```{r libraries, message=FALSE, warning=FALSE, echo = T, eval = F}
#library(tidyverse)
library(GenomicRanges)
library(rtracklayer)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
library(stringr)
library(plyranges)
#BiocManager::install("biomaRt")
library(biomaRt)
#BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)
#BiocManager::install("reader") # to remove extension from bw files names --> only installed on my own mac
#library(reader)
```


### chr19 peaks in GRanges 
```{r prepare monocyt data, echo = T, eval = F}
# H3k4me1 
mono_h3k4me1 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547981.H3K4me1.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K4me3  
mono_h3k4me3 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547984.H3K4me3.bwa.GRCh38.20150527.chr19.bed", format = "narrowPeak")
# H3K9me3 
mono_h3k9me3 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547982.H3K9me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K27ac 
mono_h3k27ac <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547980.H3K27ac.bwa.GRCh38.20150527.chr19.bed", format = "narrowPeak")
# H3K27me3   
mono_h3k27me3 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547983.H3K27me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# H3K36me3
mono_h3k36me3 <- rtracklayer::import("data/blueprint/chip_bed/C000S5H2.ERX547979.H3K36me3.bwa.GRCh38.broad.20150527.chr19.bed", format = "broadPeak")
# create list
mono_list  <- GRangesList(mono_h3k4me1, mono_h3k4me3, mono_h3k9me3, mono_h3k27ac, mono_h3k27me3, mono_h3k36me3)
# add names to each element in the list
names(mono_list) <- c("h3k4me1", "h3k4me3", "h3k9me3", "h3k27ac", "h3k27me3", "h3k36me3")
```

```{r save prepare monocyt data, echo = T, eval = F}
save(mono_h3k4me1, mono_h3k4me3, mono_h3k9me3, mono_h3k27ac, mono_h3k27me3, mono_h3k36me3, mono_list, file =  "data/prepared_rds/blueprint_monocyte_chr19_granges.RData")
```

### NOT USED - chr19 GRanges of transcript-centered genomic features      
Had the idea to use this but decided not to anymore in practical (2021 version)  
```{r prepare-anno, echo = T, eval = F}
# make my own list of GRanges objects of differnt genomic features (transcript-centered)

seqlevels(txdb) <- "chr19" 

# generate list of genomic annotations
proximal.promoter.cutoff=1000L # Specify the cutoff in bases to classify proximal promoter or enhancer. Peaks that reside within proximal.promoter.cutoff upstream from or overlap with transcription start site are classified as proximal promoters. The default is 1000 bases.
immediate.downstream.cutoff=1000L # Specify the cutoff in bases to classify immediate downstream region or enhancer region. Peaks that reside within immediate.downstream.cutoff downstream of gene end but not overlap 3 prime UTR are classified as immediate downstream. 
#' The default is 1000 bases.

transcripts_chr19 <-  unique(transcripts(txdb, columns=NULL))
exons_chr19 <- exons(txdb, columns=NULL)
introns_chr19 <- unique(unlist(intronsByTranscript(txdb)))
fiveUTRs_chr19 <- unique(unlist(fiveUTRsByTranscript(txdb)))
threeUTRs_chr19 <- unique(unlist(threeUTRsByTranscript(txdb)))
promoters_chr19 <- unique(promoters(txdb, upstream=proximal.promoter.cutoff, downstream=0))
immediateDownstream_chr19 <- unique(flank(transcripts_chr19, width=immediate.downstream.cutoff, start=FALSE, use.names=FALSE))

# make list of annotations
annotation_chr19 <- list(exons_chr19, introns_chr19, fiveUTRs_chr19, threeUTRs_chr19, promoters_chr19, immediateDownstream_chr19)

# remove all metadata information
annotation_chr19 <- lapply(annotation_chr19, function(.anno){mcols(.anno)<-NULL; .anno, echo = T, eval = F})

# add names
names(annotation_chr19)[1:6] <- c("Exons", "Introns", "fiveUTRs", "threeUTRs", "Promoters", "immediateDownstream")

# convert to GRangeslist
annotation_chr19_gr <- GRangesList(annotation_chr19)

# to make 'intergenicRegions', add up all above annotated regions and identify the remaining gaps
# we ignore strand of the annotations because peaks are unstranded
newAnno <- c(unlist(annotation_chr19_gr))
newAnno.rd <- newAnno
strand(newAnno.rd) <- "*"
newAnno.rd <- reduce(trim(newAnno.rd))
intergenicRegions_chr19 <- gaps(newAnno.rd, end=seqlengths(txdb))
intergenicRegions_chr19 <- intergenicRegions_chr19[strand(intergenicRegions_chr19)=="*"]
names(intergenicRegions_chr19) <- NULL

# add intergenicRegions to annotation list
annotation_chr19_gr$intergenicRegions <- intergenicRegions_chr19

rm(newAnno.rd, newAnno, annotation_chr19, transcripts_chr19)

# save the final GRangesList object
saveRDS(annotation_chr19_gr, file = "data/prepared_rds/txdb_annotation_chr19_grangeslist.rds")
```

Not used anymore in practical (2021 version)  
```{r prepare txdb, echo =T, echo = T, eval = F}
genes <- unique(genes(txdb, filter = list(tx_chrom = "chr19")))
transcripts<-  unique(transcripts(txdb, filter = list(tx_chrom = "chr19")))
promoters_def <- unique(promoters(txdb, filter = list(tx_chrom = "chr19"))) # upstream=2000, downstream=200
promoters_3k <- unique(promoters(txdb, filter = list(tx_chrom = "chr19"), upstream = 3000, downstream = 3000)) 

# save above GRanges objects in RData file
save(genes, transcripts, promoters_def, promoters_3k, file =  "data/prepared_rds/txdb_annotation_chr19_genes_tx_pr_tss_granges.RData")
```

### NOT USED - chr19 df of gene symbols and entrezIDs    
Not used anymore in practical (2021 version)  
```{r geneid mapping, , echo = T, echo = T, eval = F}
# map ENTREZ gene id to gene SYMBOL
entrezid_symbols <- AnnotationDbi::select(org.Hs.eg.db, keys = genes$gene_id, columns = c("ENTREZID", "SYMBOL"), keytype = "ENTREZID")
# save the final GRangesList object
saveRDS(entrezid_symbols, file = "data/prepared_rds/gene_entrezids_genesymbols_df.rds")

```

### NOT USED - chr19 df of ENCODE cis-regulatory elements    
Not used anymore in practical (2021 version)  
```{r encode cres, echo = T, eval = F}
library(plyr)
# Supplementary Table 10 from 'The ENCODE consortium, 2020, nature'
# https://www.nature.com/articles/s41586-020-2493-4#MOESM12
# table has same format as .bed file so probably 0-based
# r works 1-based, therefore have to set 'start.in.df.are.0based = T' 
encode <- makeGRangesFromDataFrame(read.delim("data/encode/encode_nature2020_supldata10_chr19.txt", stringsAsFactors = F, header = T), keep.extra.columns = T, starts.in.df.are.0based=T)
names(encode) <- encode$cCRE_accession

encode_ucsc <- rtracklayer::import("data/encode/encode_cCREs_ucscgenomebrowser_chr19.bed")
names(encode_ucsc) <- encode_ucsc$name

# add itemRBG from ucsc file to encode publication file
encode$ucsc_rgb <- encode_ucsc[match(encode$cCRE_accession, encode_ucsc$name),]$itemRgb

# check that each name as the right itemRGB
encode_labels <- arrange(as.data.frame(mcols(encode[, c("cCRE_accession", "ucsc_rgb")])), desc(cCRE_accession))
encodeucsc_labels <- arrange(as.data.frame(mcols(encode_ucsc[, c("name", "itemRgb")])), desc(name))
all(encode_labels$cCRE_accession == encodeucsc_labels$name)
all(encode_labels$ucsc_rgb == encodeucsc_labels$itemRgb)
head(encodeucsc_labels)
head(encode_labels)
all_labels <- merge(encode_labels, encodeucsc_labels, by.x = "cCRE_accession", by.y = "name", all.x = T, all.y = T)
identical(all_labels$ucsc_rgb,all_labels$itemRgb)
rm(all_labels, encode_labels, encodeucsc_labels)

# translate itemRGB into a categorie 
# ref: http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=hg38&g=encodeCcreCombined
itemRGBs <- data.frame(names(table(encode_ucsc$itemRgb)))
colnames(itemRGBs) <- "hex"
itemRGBs$hex <- as.character(itemRGBs$hex)
itemRGBs$color <- c("blue", "red", "orange", "pink", "yellow")
itemRGBs$ucsc_labels <- c("CTCF", "prom", "enhP", "K4m3", "enhD")
#itemRGBs

# add categorie from ucsc website to encode publication file
# based on reported itemRGB
encode$ucsc_label <- as.factor(encode$ucsc_rgb)
encode$ucsc_label <- as.character(mapvalues(encode$ucsc_label, from = c("#00B0F0", "#FF0000", "#FFA700", "#FFAAAA", "#FFCD00"), to = c("CTCF", "prom", "enhP", "K4m3", "enhD")))
encode
table(encode$group, encode$ucsc_label)

# make simplfied object for tutorial
encode_chr19 <- encode[, c("cCRE_accession", "ucsc_label")]
show(encode_chr19)
# save this object
saveRDS(encode_chr19, file = "data/prepared_rds/encode_ccres_chr19.rds")
```

### chr19 BLUEPRINT monocytes transcript quantification  

- add entrez gene id using org.Hs.eg.db package   
- restrict to genes on chromosome 19  

```{r quantification, echo = T, eval = F}
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)

all_results <- read.table("data/blueprint/rna_quantification/C000S5B1.gene_quantification.rsem_grape2_crg.GRCh38.20150622.results", header =T)

# remove version from ensemble gene id 
all_results$gene_id <- sapply(all_results$gene_id,  sub, pattern = "\\.\\d+$", replacement = "")

# map entrez id to ensemble id
all_results$entrezgene_id = mapIds(org.Hs.eg.db,
                    keys=all_results$gene_id, 
                    column="ENTREZID",
                    keytype="ENSEMBL",
                    multiVals="first")

# vector of entrez ids for genes on chr19 
genes_chr19 <- genes(txdb)
geneids_genes_chr19 <- mcols(genes_chr19)[,1]

# subset all_results for genes on chr19
quantification_chr19 <- all_results[all_results$entrezgene_id %in% geneids_genes_chr19,c("entrezgene_id", "TPM", "FPKM")]
```

```{r save quantification, echo = T, eval = F}
# write quantification_chr19 to rds
saveRDS(quantification_chr19, file = "data/prepared_rds/blueprint_c000s5_gene_quantification_chr19.rds")

```

#### NOT USED - chr19 Alternative gene ID  mapping with biomaRt:  

- add entrez gene id using biomaRt (see steps on )  
- restrict to genes on chromosome 19  

Ref: https://www.biostars.org/p/302441/#302485  

```{r quantification biomart, echo = T, eval = F}
all_results <- read.table("data/blueprint/rna_quantification/C000S5B1.gene_quantification.rsem_grape2_crg.GRCh38.20150622.results", header =T)

# remove version from ensemble gene id  with stringi
all_results$gene_id_short  <- str_replace(all_results$gene_id,
                        pattern = ".[0-9]+$",
                        replacement = "")
# obtain mart object 
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")



# Test example from https://www.biostars.org/p/302441/#302485  
gene_ids_version <- c("ENSG00000236246.1", # note: none of these can be mapped to entrezgene_id
                      "ENSG00000281088.1",
                      "ENSG00000254526.1",
                      "ENSG00000223575.2",
                      "ENSG00000201444.1",
                      "ENSG00000232573.1")
gene_ids <- str_replace(gene_ids_version,
                        pattern = ".[0-9]+$",
                        replacement = "")

getBM(attributes = c('ensembl_gene_id',
                     'entrezgene_id'),
      filters = 'ensembl_gene_id', 
      values = gene_ids,
      mart = mart)

getBM(attributes = c('ensembl_gene_id',
                     'entrezgene_id'),
      filters = 'ensembl_gene_id', 
      values = c('ENSG00000001460', 'ENSG00000008517', 'ENSG00000009724'), #these should map
      mart = mart)

# map entrez id to ensemble id with biomaRt 
#library(biomaRt)
geneids <- all_results$gene_id_short

mappings <- getBM(attributes = c('ensembl_gene_id','entrezgene_id'), 
      values = geneids,
      filters = 'ensembl_gene_id', 
      mart = mart)

dim(mappings)

all_results <- merge(all_results, mappings, by.x = "gene_id_short" , by.y = "ensembl_gene_id")

# vector of entrez ids for genes on chr19 
genes_chr19 <- genes(txdb)
geneids_genes_chr19 <- mcols(genes_chr19)[,1]

# subset all_results for genes on chr19
quantification_chr19_biomart <- all_results[all_results$entrezgene_id %in% geneids_genes_chr19,]


```


### chr19 Complex Heatmap object week 2  
Used for exercises in fg2 (2021 version)  
This function is also run by students but resulting object by several exercises and as the function takes a while, I decided to prepare the object in advance.  
```{r, echo = T, eval = F}
# prepare objects for upset plot in week 2
monocytes_combinationmatrix <- make_comb_mat(mono_list) # takes a while! 

saveRDS(monocytes_combinationmatrix, file =  "data/prepared_rds/blueprint_monocyte_chr19_complexheatmap_combmatrix.rds")

```

### NOT USED - Select bigwig files for chr19 limited to regions TSS +/- 20kb   
**2020-01-17**  
This is to reduce the .bw file size as for tutorial we should have files of Â±5MB and useing onlu chr19 still leaves 20-30MB per file. 
I've tried a w of 40004, 20004, 15002; but output remains > 10 MB while.  
  
40004 - 16.7 MB  
20004 - 13.7 MB   
15002 - 12.5 MB    
10002 - 11 MB  
<br>
Decide to check out other functions (on the server). If that doesn't work, talk with Siebren.  
```{r select regions in bigwig files, eval = F}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
seqlevels(txdb) <- seqlevels0(txdb)
seqlevels(txdb) <- "chr19"

w <- 10002

tss_chr19 <- unique(promoters(genes(txdb), upstream = 2000, downstream = 0 )) 
length(tss_chr19)
tss_chr19 <- resize(tss_chr19, width = 1, fix = "end")
tss_chr19_window <- resize(tss_chr19, width = w, fix = "center")
tss_chr19_window_reduce <- reduce(tss_chr19_window)
sum(width(tss_chr19_window_reduce))
blueprint_file_dir <- "/Users/christatoenhake/Documents/Werk/16_moldevbio_onderwijsvernieuwing/fg1_bb064b_new_practicum/fg1_r_data/blueprint/chip_bw/week3"

bw_h3k4me3_tss_chr19_window <- import("/Users/christatoenhake/Documents/Werk/16_moldevbio_onderwijsvernieuwing/fg1_bb064b_new_practicum/fg1_r_data/blueprint/chip_bw/week3/C000S5H2.ERX547984.H3K4me3.bwa.GRCh38.20150528.chr19.bw", which = tss_chr19_window_reduce, as  = "RleList")


bw_out <- file.path(blueprint_file_dir, "test_out.bw")

export.bw(object = bw_h3k4me3_tss_chr19_window, bw_out )


```
<br>
### NOT USED - chr21 - Test bigwig size of chr21 TSS windows
What if I used chr21?  
<br>
40004 - 3.1 MB  
20004 - 2.6 MB   
15002 - 2.4 MB    
10002 - 0.5 MB  
<br>
**These files are small enough! :-D**   

> **However: later found out that FPKM per quantile does not look representative for genome wide picture.**  

```{r wk3 test tss chr21, warning=F, eval = F}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
seqlevels(txdb) <- seqlevels0(txdb)
seqlevels(txdb) <- "chr21"
tss_chr21 <- unique(promoters(txdb, upstream = 2000, downstream = 0 )) 
tss_chr21 <- resize(tss_chr21, width = 1, fix = "end")
  
blueprint_file_raw <- "/Users/christatoenhake/Documents/Werk/16_moldevbio_onderwijsvernieuwing/fg1_bb064b_new_practicum/fg1_r_data/blueprint/chip_bw"

blueprint_file_chr21_tss <- "/Users/christatoenhake/Documents/Werk/16_moldevbio_onderwijsvernieuwing/fg1_bb064b_new_practicum/fg1_r_data/blueprint/chip_bw/wk3_chr21_tss_window/"
  
bw_files <- dir(path = blueprint_file_raw, pattern = "*.bw")

resize_bw <- function(w_in_kb, f, tss){
  print(deparse(substitute(tss)))
  w_in_bp <- w_in_kb*1000+4
  tss_window <- resize(tss, width = w_in_bp, fix = "center")
  tss_window_reduce <- reduce(tss_chr21_window)
  sum(width(tss_window_reduce))
  bw_h3k4me3_tss_chr21_window <- import(paste(blueprint_file_raw, f, sep = "/"), which = tss_window_reduce, as = "RleList")
  bw_out <- file.path(blueprint_file_chr21_tss, 
                      paste(rmv.ext(f, only.known = F), "chr21_TSS", w_in_kb, "kb.bw", sep = "_"))
  export.bw(object = bw_h3k4me3_tss_chr21_window, bw_out)
}
for(file in bw_files){
resize_bw(40, file, tss_chr21)
resize_bw(20, file, tss_chr21)
}
```
<br>
And what if we add regions around DNaseI peaks?  
<br>
Take 'peaks' files, which contains the "Peaks are 150nt regions that overlap or are contained within hotspots, which are areas of very high density relative to background.".  
```{r wk3 test chr21 tss plus DNaseI, warning=F, eval = F}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
seqlevels(txdb) <- seqlevels0(txdb)
seqlevels(txdb) <- "chr21"
tss_chr21 <- unique(promoters(txdb, upstream = 2000, downstream = 0 )) 
tss_chr21 <- resize(tss_chr21, width = 1, fix = "end")

dnase <- import.bed("/Users/christatoenhake/Documents/Werk/16_moldevbio_onderwijsvernieuwing/fg1_bb064b_new_practicum/fg1_r_data/blueprint/chip_bed/C004084E.ERX197157.Dnase.GRCh38.hotspot.peaks.20150709.bed")
dnase_chr21 <- dnase %>% plyranges::filter(seqnames == "chr21")

blueprint_file_raw <- "/Users/christatoenhake/Documents/Werk/16_moldevbio_onderwijsvernieuwing/fg1_bb064b_new_practicum/fg1_r_data/blueprint/chip_bw"

blueprint_file_chr21_tss <- "/Users/christatoenhake/Documents/Werk/16_moldevbio_onderwijsvernieuwing/fg1_bb064b_new_practicum/fg1_r_data/blueprint/chip_bw/wk3_chr21_tss_dnase_window/"
  
bw_files <- dir(path = blueprint_file_raw, pattern = "*.bw")
w_in_kb <- 40
resize_bw_tss_dnase <- function(w_in_kb, f, tss, dnase_peaks){
  #print(deparse(substitute(tss)))
  w_in_bp <- w_in_kb*1000+4
  tss_window <- resize(tss_chr21, width = w_in_bp, fix = "center")
  dnase_window <- resize(dnase_chr21, width = w_in_bp, fix = "center")
  tss_dnase_window <- c(tss_window, dnase_window)
  tss_dnase_window_reduce <- reduce(tss_dnase_window)
  bw_h3k4me3_tss_dnase_chr21_window <- import(paste(blueprint_file_raw, f, sep = "/"), which = tss_dnase_window_reduce, as = "RleList")
  bw_out <- file.path(blueprint_file_chr21_tss, 
                      paste(rmv.ext(f, only.known = F), "chr21_TSS_DNasePeaks", w_in_kb, "kb.bw", sep = "_"))
  export.bw(object = bw_h3k4me3_tss_dnase_chr21_window, bw_out)
}
for(file in bw_files){
resize_bw_tss_dnase(40, file, tss_chr21, dnase_chr21)
resize_bw_tss_dnase(20, file, tss_chr21, dnase_chr21)
}
```

## NOT USED - chr21 peaks and quantification for week 3  
 
> **Although bigwig of chip-seq data limited to chromosome 21 is nicely small enough in size. if we next visualize gene epxression per quantile this gives quantiles where the median expression of the first two quantiles is near 0. This is not representative for genome-wide expression. We therefor decide to stick with chr19 but randomly select 50% (or a bit less) of the genes per quantile to limit the data size.**  

#### NOT USED - chr21 peaks in GRanges  
Note that I deleted the "chr21_bed" folder in "inst/extdata/week3/" after running this chunk to save space in the learnr.proto library. This is also why `eval=F` is set.    
The files in this folder can be found on cn54: `scratch/ctoenhake/data/fg1_r_data/blueprint/chip_bed`. Whole genome bed peak files were downloaded from blueprint data portal > Experiments > search for C000S5.   
```{r prepare monocyt data, echo = T, eval = F}
# H3k4me1 
monocytes_h3k4me1 <- rtracklayer::import("inst/extdata/week3/chr21_bed/C000S5H2.ERX547981.H3K4me1.bwa.GRCh38.broad.20150527.chr21.bed", format = "broadPeak")
# H3K4me3  
monocytes_h3k4me3 <- rtracklayer::import("inst/extdata/week3/chr21_bed/C000S5H2.ERX547984.H3K4me3.bwa.GRCh38.20150527.chr21.bed", format = "narrowPeak")
# H3K9me3 
monocytes_h3k9me3 <- rtracklayer::import("inst/extdata/week3/chr21_bed/C000S5H2.ERX547982.H3K9me3.bwa.GRCh38.broad.20150527.chr21.bed", format = "broadPeak")
# H3K27ac 
monocytes_h3k27ac <- rtracklayer::import("inst/extdata/week3/chr21_bed/C000S5H2.ERX547980.H3K27ac.bwa.GRCh38.20150527.chr21.bed", format = "narrowPeak")
# H3K27me3   
monocytes_h3k27me3 <- rtracklayer::import("inst/extdata/week3/chr21_bed/C000S5H2.ERX547983.H3K27me3.bwa.GRCh38.broad.20150527.chr21.bed", format = "broadPeak")
# H3K36me3
monocytes_h3k36me3 <- rtracklayer::import("inst/extdata/week3/chr21_bed/C000S5H2.ERX547979.H3K36me3.bwa.GRCh38.broad.20150527.chr21.bed", format = "broadPeak")
# create list
monocytes_list  <- GRangesList(monocytes_h3k4me1, monocytes_h3k4me3, monocytes_h3k9me3, monocytes_h3k27ac, monocytes_h3k27me3, monocytes_h3k36me3)
# add names to each element in the list
names(monocytes_list) <- c("h3k4me1", "h3k4me3", "h3k9me3", "h3k27ac", "h3k27me3", "h3k36me3")
# make 1 GRanges object with all ChIPs
monocytes_all <- unlist(monocytes_list)
monocytes_all$chip <- names(monocytes_all)
```

#### NOT USED - chr21 quantification   
Note that I deleted the "quant" folder in "inst/extdata/week3/" after running this chunk because of the size of the file in it ("C000S5B1.gene_quantification.rsem_grape2_crg.GRCh38.20150622.results"). This is also why `eval=F` is set.  
This was downloaded from http://dcc.blueprint-epigenome.eu/#/experiments/ERX157053 --> "Transcription quantification (Genes)" file.  

```{r quantification, echo = T, eval = F}
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)

all_results <- read.table("inst/extdata/week3/quant/C000S5B1.gene_quantification.rsem_grape2_crg.GRCh38.20150622.results", header =T)

# remove version from ensemble gene id 
all_results$gene_id <- sapply(all_results$gene_id,  sub, pattern = "\\.\\d+$", replacement = "")

# map entrez id to ensemble id
all_results$entrezgene_id = mapIds(org.Hs.eg.db,
                    keys=all_results$gene_id, 
                    column="ENTREZID",
                    keytype="ENSEMBL",
                    multiVals="first")

# set sqlevels in txdb to chr21
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
seqlevels(txdb) <- seqlevels0(txdb)
seqlevels(txdb) <- "chr21"

# vector of entrez ids for genes on chr21 
genes_chr21 <- genes(txdb)
geneids_genes_chr21 <- mcols(genes_chr21)[,1]

# subset all_results for genes on chr21
quantification_chr21 <- all_results[all_results$entrezgene_id %in% geneids_genes_chr21,c("entrezgene_id", "TPM", "FPKM")]
```

#### NOT USED - store chr21 granges and quantification  
```{r save prepare monocyt data, echo = T, eval = F}
save(monocytes_h3k4me1, monocytes_h3k4me3, monocytes_h3k9me3, monocytes_h3k27ac, monocytes_h3k27me3, monocytes_h3k36me3, monocytes_list, monocytes_all, quantification_chr21, file =  "inst/extdata/week3.Rdata")
```

## chr19 - get smaller bigwig by downsampling genes   
Problem: chip-seq bigwig files of 1 chromosome are too large  
Aim: for 5MB per tutorial on data.  
Possible solutions:  

1. Select chr19, only 20kb +/- TSS sites \-\-\> files remain too large   
2. Work with chr21 \-\-\> files are small enough but gene expression is not representative for genome wide  
3. Random select subsample of genes from chr19 based on quantification quartiles and used these to select regions in bigwig file.    

```{r step 1 chr19 subsample per gene expression quantile, eval = F}
# Important parameter: the chosen proportion of genes
chosen_proportion <- 0.4
```

```{r step 2 chr19 subsample per gene expression quantile, eval = F}
### set sqlevels in txdb to chr19
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene 
seqlevels(txdb) <- seqlevels0(txdb)
seqlevels(txdb) <- "chr19"

# load chr19 quantification (used in fg2)
rdsfile <- system.file("extdata", "week2", "prepared_rds", "blueprint_c000s5_gene_quantification_chr19.rds", package = "learnr.proto")
quantification_chr19 <- readRDS(rdsfile)
```

```{r step 3 chr19 subsample per gene expression quantile, eval = F}
# define new variable based on quartiles of gene expression
quantification_chr19 <- quantification_chr19 %>% 
  mutate(expression_group = ntile(FPKM, 4))

### set seed for slice_sample
set.seed(13)

# subsample per quartile  
quantification_chr19_sample <- quantification_chr19 %>% 
  group_by(expression_group) %>% 
  slice_sample(prop=chosen_proportion, replace = F)

# check number of genes per quantile and in expression plot
table(quantification_chr19_sample$expression_group)
ggplot(quantification_chr19, aes(x = as.factor(expression_group), y = log10(FPKM+1)))+
  geom_boxplot()+
  geom_jitter()
ggplot(quantification_chr19_sample, aes(x = as.factor(expression_group), y = log10(FPKM+1)))+
  geom_boxplot()+
  geom_jitter()
```

```{r setp 4 subsample bigiwg file for these TSS windows, eval = F}
# subsample bigwig for regions around the TSS of the subsampled genes on chr19  
## set fref files and libraries:
tss_chr19 <- unique(promoters(genes(txdb), upstream = 2000, downstream = 0 )) 
tss_chr19 <- resize(tss_chr19, width = 1, fix = "end")
tss_chr19_sample <- tss_chr19 %>% filter(gene_id %in% quantification_chr19_sample$entrezgene_id )  

### input dir:
dir_raw_files <- "/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19/"
### output dir:
dir_out_files <- "/home/ctoenhake/learnr.proto/inst/extdata/week3/bw_chr19_tss_sample/"

### function to select regions in bigwig files and write output in outdir
bw_files <- dir(path = dir_raw_files, pattern = "*.bw")
resize_bw <- function(w_in_kb, f, tss){
  print(deparse(substitute(f)))
  w_in_bp <- w_in_kb*1000+4
  tss_window <- resize(tss, width = w_in_bp, fix = "center")
  sum(width(tss_window))
  bw_tss_window <- import(paste(dir_raw_files, f, sep = "/"), which = tss_window, as = "RleList")
  bw_out <- file.path(dir_out_files, 
                      paste0(tools::file_path_sans_ext(f), "_chr19tss_", chosen_proportion, "ofGenes_", w_in_kb, "kb.bw"))
  export.bw(object = bw_tss_window, bw_out)
  rm(tss_window, bw_tss_window, bw_out)
}

## perform function on all files in input directory. 
for(file in bw_files){
resize_bw(15, file, tss_chr19_sample)
}

## save sliced quantification and TSSs
save(quantification_chr19_sample, tss_chr19_sample, file = "/home/ctoenhake/learnr.proto/inst/extdata/week3.Rdata")
```

### session Info
```{r session info, echo = T, eval = F}
sessionInfo()

```

