<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<meta name="progressive" content="false" />
<meta name="allow-skip" content="true" />

<title>Week 2: Exploratory &amp; Overlap analysis</title>


<!-- highlightjs -->
<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>

<!-- taken from https://github.com/rstudio/rmarkdown/blob/67b7f5fc779e4cfdfd0f021d3d7745b6b6e17149/inst/rmd/h/default.html#L296-L362 -->
<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("section-TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>
<!-- end tabsets -->


<link rel="stylesheet" href="css/custom.css" type="text/css" />

</head>

<body>



<div class="pageContent band">
<div class="bandContent page">

<div class="topics">

<div class="tutorial-exercise-support" data-label="prep-txdbindiv" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># this is a preperatory chunk add &#39;exercise.setup = &quot;prep-txdbindiv&quot;` to chunk header to make object available
# This .RData file imports GRanges objects with genomic locations of genes, transcripts, TSSs and promoters.
# Promoters come in two flavors:  
# 1. promoters_def = default settings from promoters() function with 2kb up and 0.2 kb downstream the TSS
# 2. promoters_chr19_3k = for heatmaps, with 3kb upstream and 3kb downstream the TSS
load(paste0(datadir, &quot;www/prepared_rds/txdb_annotation_chr19_genes_tx_pr_tss_granges.RData&quot;))</code></pre>
</div>
<div id="section-introduction" class="section level2">
<h2>2.1 Introduction</h2>
<p><br> In week 1 we discussed the major steps imvolved in obtaining genomics data, from experiment to raw data to normalized signal and peaks. We examined histone PTM ChIP-, DNaseI-, and RNA-seq data of monocytes in the UCSC genome browser and searched for regions with increased signal, <em>peaks</em>. We looked at the position of these peaks in the chromosome and with respect to genes as well as the co-occurence of different marks.<br />
<br> These observations showed you that some marks are associated with particular functional elements (eg active enhancer, active promoter, silenced domains).<br />
<br> <strong>This week and next week we will put numbers to these observations and perform computational analyses to answer common questions including:</strong> “<em>How many peaks do I have?</em>”, “<em>Is this mark statistically enriched in a particular genomic element?</em>”, “<em>Which and how often do marks co-occur?</em>”,“<em>The expression of which gene could be affected by this mark?</em>”, and “<em>What is the signal of the mark around a particular element of interest like the TSS?</em>”<br />
<br></p>
<div id="section-learning-objectives" class="section level3">
<h3>2.1.1 Learning Objectives</h3>
<blockquote>
<p>At the end of week 2 you are able to:</p>
<ol style="list-style-type: decimal">
<li>Import ChIP-seq peaks into a GRanges object in r.<br />
</li>
<li>Perform exploratory data analysis on GRanges objects with ChIP-seq peaks.<br />
</li>
<li>Detect and count overlap between two GRanges objects.<br />
</li>
<li>Visualize the number of overlaps in a vennDiagram.<br />
</li>
<li>Visualize the number of overlaps with bar- and pieplots.<br />
</li>
<li>Statistically test for enrichment of histone marks in a particular genomic region.</li>
<li>Optionally, quantify and plot the overlap among histone marks (a tiny bit like chromHMM but much less fancy, ggplot + geom_bar with index at the bottom for the different combinations or plot the tabular output as heatmap with nubmer display).<br />
</li>
<li>Identify the nearest TSS to a histone ChIP-seq peak.<br />
</li>
<li>Identify the genes that are associated with these TSSs.<br />
These refer to global learning objectives #4-#7.</li>
</ol>
</blockquote>
<p><br></p>
<hr />
<p><br></p>
<blockquote>
<p>TEASER… let’s look at some of the plots you will make.</p>
</blockquote>
<p><br></p>
<hr />
<p><br></p>
</div>
</div>
<div id="section-chip-seq-peaks-in-r" class="section level2" data-progressive="TRUE">
<h2 data-progressive="TRUE">2.2 ChIP-seq peaks in R</h2>
<div id="section-data-import" class="section level3">
<h3>2.2.1 Data import</h3>
<p>In genomics we often work with <strong>interval</strong> data. Think of peaks, genes, exons, … any genomic region reported with the <strong>genomic coordinates: chr, start, and end</strong>. These often come in a tabular format with the basic information about the location and some other information. Common file formats include BED, GFF or BigWig files.<br />
<br> Peak data can be imported into a data.frame in r with:</p>
<pre class="r"><code>object_name &lt;- read.table(&quot;location/of/peak_file.bed&quot;)</code></pre>
<p><br> The data.frame datastructure is, however, not the most efficient way to work with interval data. E.g. a simple manipulation such as shifting all reported intervals 2 bp to the right, requires you to manipulate the “start” and “end” columns at the same time.<br />
<br> Interval data such as genomic peaks can be more efficiently handled with the IRanges package which works with a data structure especially developed for <strong>ranges of integers</strong>: IRanges objects.<br />
<br> To construct an IRanges object we require at least two of the following three values:</p>
<ol style="list-style-type: decimal">
<li>a starting coordinate<br />
</li>
<li>a finishing coordinate<br />
</li>
<li>the width of the interval.</li>
</ol>
<p><br> The GRanges objects of the GenomicRanges package are very similar but require a additional <strong>sequence name</strong> (in other words a chromosome) for every interval and an optional <strong>strand</strong> column.<br />
<br> Here’s an example GRanges object called <code>my_gr</code> with the functions to set and retrieve its content:</p>
<p><img src="images/wk2_2_GRanges_fields_my_gr.png" style="width:100.0%" /></p>
<p><br> Notice that this object holds the <strong>genomic coordinates</strong> of the peaks and <strong>2 metadata columns</strong>. Metadata is <strong>data about the data</strong>:</p>
<table>
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Genomic coordinates</th>
<th align="left">Metadata columns</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Printed on the left-hand side of the</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Extract using <code>granges(object_name)</code></td>
<td align="left">Extract as DataFrame with <code>mcols(object_name)</code> or <code>object_name$column_name</code> for a specific column</td>
</tr>
<tr class="odd">
<td align="left">Columns are restricted to <code>seqnames</code>, <code>ranges</code> and <code>strand</code></td>
<td align="left">Almost anything can be stored in the metadata</td>
</tr>
</tbody>
</table>
<p><br> Information about the genome is stored in the <code>seqinfo</code> part of the object. In the example you see that the peaks are located in human genome 38 which holds 455 sequences in total of which 1 circular (mitochondrial genome).<br />
<br> GRanges objects follow the <strong>tidy data principle</strong>: each row of a Ranges object corresponds to an interval, and each column will represent a variable about that interval, and generally each object will represent a single unit of an observation (like gene annotations).<br />
<br> We can use <strong>dplyr</strong>-like functions to manipulate these objects and use the <strong>pipe</strong> operator <code>%&gt;%</code> to combine functions in a workflow. A brief reminder of common dplyr-functions and examples of plyranges-relatives on GRanges objects:</p>
<table>
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Function</th>
<th align="left">Purpose</th>
<th align="left">Example on GRanges</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><code>select()</code></td>
<td align="left">subset variables (=columns)</td>
<td align="left"><code>select(my_gr, GC)</code></td>
</tr>
<tr class="even">
<td align="left"><code>group_by()</code></td>
<td align="left">group data into rows with the same value for the specified variable.</td>
<td align="left"><code>my_gr %&gt;% group_by(strand)</code>.</td>
</tr>
<tr class="odd">
<td align="left"><code>filter()</code></td>
<td align="left">subset observations (= rows)</td>
<td align="left"><code>filter(my_gr, GC &lt; 0.4)</code> or <code>my_gr %&gt;% group_by(strand) %&gt;%  filter(GC == max(GC))</code>.</td>
</tr>
<tr class="even">
<td align="left"><code>summarize()</code></td>
<td align="left">Summarise variables, often per group</td>
<td align="left"><code>group_by(my_gr, strand) %&gt;% summarize(n = n(), gc = max(GC))</code></td>
</tr>
</tbody>
</table>
<p><br><br />
To parse peak files directly into GRanges we use the <code>import()</code> function from the rtracklayer package:</p>
<pre class="r"><code># import monocyte H3K4me3 peak locations
mono_h3k4me3 &lt;- rtracklayer::import(paste0(datadir, &quot;www/blueprint/bed/C000S5H2.ERX547984.H3K4me3.bwa.GRCh38.20150527.chr19.bed&quot;), format = &quot;narrowPeak&quot;)  </code></pre>
<p><br></p>
<blockquote>
<p><em>Background</em>: you will come across <strong>narrowPeak</strong> and <strong>broadPeak</strong> files. These are both outputs from MACS2 peak calling.</p>
<ul>
<li>narrowPeak files are for more focused signals like transcription factor ChIP-seq but also some histone marks.<br />
</li>
<li>broadPeak mode is used for more defuse signals.</li>
</ul>
<p>The outputs are very similar except that the broadPeak mode, the peak summit won’t be called and the values in the 5th, and 7-9th columns are the mean value across all positions in the peak region.</p>
<table>
<thead>
<tr class="header">
<th align="left">narrowPeak file</th>
<th align="left">broadPeak file</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">H3K27ac</td>
<td align="left">H3K27me3</td>
</tr>
<tr class="even">
<td align="left">h3k4me3</td>
<td align="left">H3K36me3</td>
</tr>
<tr class="odd">
<td align="left">H3K9/14ac</td>
<td align="left">H3K9me3</td>
</tr>
<tr class="even">
<td align="left">H2A.Zac</td>
<td align="left">H3K4me1</td>
</tr>
</tbody>
</table>
</blockquote>
<p>Let’s take a quick look at one of these files.</p>
<blockquote>
<p><strong>Exercise 1:</strong><br />
Print the head 2 lines of <code>mono_h3k4me3</code>.</p>
</blockquote>
<div class="tutorial-exercise-support" data-label="q1_head_h3k4me3-setup" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>mono_h3k4me3 &lt;- rtracklayer::import(paste0(datadir, &quot;www/blueprint/bed/C000S5H2.ERX547984.h3k4me3.bwa.GRCh38.20150527.chr19.bed&quot;), format = &quot;narrowPeak&quot;)</code></pre>
</div>
<div class="tutorial-exercise" data-label="q1_head_h3k4me3" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div id="section-q1_head_h3k4me3-hint">
<p><strong>Hint:</strong> You may want to use the <code>head(.., n = ..)</code> function.</p>
</div>
<div class="tutorial-exercise-support" data-label="q1_head_h3k4me3-solution" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>head(mono_h3k4me3, n = 2)</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="q1_head_h3k4me3-code-check" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>grade_code()</code></pre>
</div>
<blockquote>
<p><em>Note 1</em>: We have restricted all peak files to chromosome 19 to limit their file size. That is why “seqinfo” holds only one sequence.</p>
</blockquote>
<blockquote>
<p><em>Note 2</em>: The original .BED files lacked column headers. The <code>rtracklayer::import()</code> function appendended variable names and converted the data to the data type (eg numeric, integer, character etc.) according to the file format we defined.</p>
</blockquote>
<p>We have the following information about the identified peaks (ref: <a href="https://github.com/macs3-project/MACS">MACS2 github</a>):</p>
<table>
<colgroup>
<col width="25%" />
<col width="25%" />
<col width="25%" />
<col width="25%" />
</colgroup>
<thead>
<tr class="header">
<th>Column</th>
<th>Information</th>
<th>narrowPeak description</th>
<th>broadPeak description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1.</td>
<td>chromosome name</td>
<td>same</td>
<td>same</td>
</tr>
<tr class="even">
<td>2.</td>
<td>peak start</td>
<td>same</td>
<td>same</td>
</tr>
<tr class="odd">
<td>3.</td>
<td>peak end</td>
<td>same</td>
<td>same</td>
</tr>
<tr class="even">
<td>4.</td>
<td>peak name</td>
<td>same</td>
<td>same</td>
</tr>
<tr class="odd">
<td>5.</td>
<td>score</td>
<td>-10log(qvalue) * 10, rounded down to integer value)</td>
<td>mean of -10log(qvalue) * 10 <em>across all positions</em></td>
</tr>
<tr class="even">
<td>6.</td>
<td>strand</td>
<td>+, - or * for unstranded</td>
<td>same</td>
</tr>
<tr class="odd">
<td>7.</td>
<td>fold_enrichment</td>
<td>at peak summit</td>
<td>mean across all positions</td>
</tr>
<tr class="even">
<td>8.</td>
<td>-log10(qvalue) (<em>e.g.if qvalue = 1e-10, this value is 10</em>)</td>
<td>at peak summit</td>
<td>mean <em>across all positions</em></td>
</tr>
<tr class="odd">
<td>9.</td>
<td>-log10(pvalue)</td>
<td>at peak summit</td>
<td>mean <em>across all positions</em></td>
</tr>
<tr class="even">
<td>10.</td>
<td>relative summit</td>
<td>position relative to peak start</td>
<td>not reported</td>
</tr>
</tbody>
</table>
<p><br> <strong>Let’s do some exploratory analyses on these data.</strong><br />
<br></p>
</div>
<div id="section-exploratory-data-analysis" class="section level3">
<h3>2.2.2 Exploratory data analysis</h3>
<p>A common first step in genomics data analysis is to make some exploratory plots. We will do so for <code>mono_h3k4me3</code>.<br />
<br></p>
<blockquote>
<p><strong>Exercise 2:</strong></p>
<ul>
<li>Determine the number of h3k4me3 peaks in monocytes.<br />
</li>
<li>What is the distribution of peak sizes and scores?</li>
</ul>
</blockquote>
<div class="tutorial-exercise" data-label="q2a_explore" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># How many peaks do we have?  </code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div id="section-q2a_explore-hint">
<p><strong>Hint:</strong> You may want to use the <code>length()</code> function.</p>
</div>
<pre><code>## [1] 2838</code></pre>
<div class="tutorial-exercise" data-label="q2b_explore" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># What is the distribution of peak sizes?   
summary(...(...))</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div id="section-q2b_explore-hint">
<p><strong>Hint:</strong> You may want to retrieve peak widths with <code>width()</code> and subsequently use the base <code>summary</code> function.</p>
</div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   200.0   389.2   879.5  1198.6  1661.2 12705.0</code></pre>
<div class="tutorial-exercise" data-label="q2c_explore" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># What is the distribution of peak scores?</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div id="section-q2c_explore-hint">
<p><strong>Hint:</strong> You may want to use <code>score()</code> to retrieve the peak scores.</p>
</div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##      15      84     374    1212    2415    5228</code></pre>
<p><br></p>
</div>
<div id="section-exploratory-plotting" class="section level3">
<h3>2.2.3 Exploratory plotting</h3>
<p>Let’s plot some of the above generated outputs using <strong>base</strong> plotting functions.</p>
<blockquote>
<p><strong>Exercise 3:</strong><br />
Plot the h3k4me3 peaksizes as a histogram and as a boxplot.</p>
</blockquote>
<div class="tutorial-exercise" data-label="q3a_exploreplot" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># plot the distribution of peak sizes for h3k4me3 as histogram
..(..(mono_h3k4me3), main = &quot;Monocytes h3k4me3 ChIP-seq, chr19&quot;, xlab = &quot;peak size&quot;, col = &quot;gray&quot;)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div id="section-q3a_exploreplot-hint">
<p><strong>Hint:</strong> You may want to use <code>width()</code> to retrieve the peak widths and plot with <code>hist()</code></p>
</div>
<p><img src="FunctGenWeek2_v2_files/figure-html/q3a_exploreplot-solution-1.png" width="624" /></p>
<div class="tutorial-exercise" data-label="q3b_exploreplot" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># plot the same data as boxplot
..(..., main = &quot;Monocytes h3k4me3 ChIP-seq, chr19&quot;, ... = &quot;peak size&quot;, col = &quot;gray&quot;)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div id="section-q3b_exploreplot-hint">
<p><strong>Hint:</strong> You may want to plot with <code>boxplot()</code>. Which axis will hold the widht information, x or y?</p>
</div>
<p><img src="FunctGenWeek2_v2_files/figure-html/q3b_exploreplot-solution-1.png" width="624" /></p>
<p><br></p>
<blockquote>
<p><em>Note</em>: plotting with the ggplots packages is also possible, as long as we convert the data we are interested to <code>data.frame</code> objects by running for example:</p>
</blockquote>
<pre class="r"><code>meta_h3k4me3 &lt;- as.data.frame(mcols(mono_h3k4me3))</code></pre>
<p><br> Several peaks have a width of &gt;10000 bp. You can retrieve these peaks with: <code>mono_h3k4me3[width(mono_h3k4me3)&gt; 10000,]</code> or using <code>filter()</code> from <code>plyranges</code>.<br />
<br></p>
<blockquote>
<p><strong>Exercise 4:</strong><br />
What is the number of peaks with width &gt; 10kb.</p>
</blockquote>
<div class="tutorial-exercise" data-label="q4_filter" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># How many mono_h3k4me3 peaks have width &gt; 10kb&gt;</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div id="section-q4_filter-hint">
<p><strong>Hint:</strong> You may want to use <code>[ ]</code>-subsetting and <code>width &gt; 10000</code></p>
</div>
<div id="section-q4_filter-hint">
<p><strong>Hint:</strong> You may want to use <code>length(..[..(..) &gt; 10000,])</code></p>
</div>
<pre><code>## [1] 4</code></pre>
<blockquote>
<p><strong>Exercise 5:</strong><br />
Filter <code>mono_h3k4me3</code> for peaks with width &gt; 10kb using <code>filter()</code> from <code>plyranges</code></p>
</blockquote>
<div class="tutorial-exercise" data-label="q5_filter" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># filter mono_h3k4me3 for peaks with size above 10kb
.. %&gt;% plyranges::..(.. &gt; ..)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<pre><code>## GRanges object with 4 ranges and 6 metadata columns:
##       seqnames            ranges strand |                            name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |                     &lt;character&gt;
##   [1]    chr19   4057708-4069982      * | 3337.macs2_peak_call_peak_22766
##   [2]    chr19 13094389-13105943      * | 3337.macs2_peak_call_peak_23309
##   [3]    chr19 35896740-35909444      * | 3337.macs2_peak_call_peak_24016
##   [4]    chr19 39404035-39414347      * | 3337.macs2_peak_call_peak_24149
##           score signalValue    pValue    qValue      peak
##       &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;
##   [1]      4368   100.87201 441.68515 436.88461      8901
##   [2]      3267    79.78811  330.1217 326.78381      8268
##   [3]      3088     46.5275 312.01044 308.81287     11185
##   [4]      4273    97.90211 431.89682  427.3089      4278
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><br></p>
<blockquote>
<p><strong>Exercise 6</strong><br />
Use the <code>plyranges::filter</code> and the UCSC genome browser to find the gene(s) whose promoter(s) is/are covered by the peak with the highest score.</p>
</blockquote>
<div class="tutorial-exercise" data-label="q6_maxscore" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># find the peak peak with highest score</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div id="section-q6_maxscore-hint">
<p><strong>Hint:</strong> You may want to use <code>score == max(score)</code></p>
</div>
<pre><code>## GRanges object with 1 range and 6 metadata columns:
##       seqnames            ranges strand |                            name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |                     &lt;character&gt;
##   [1]    chr19 15377793-15380841      * | 3337.macs2_peak_call_peak_23451
##           score signalValue    pValue    qValue      peak
##       &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;
##   [1]      5228     116.995  530.1864 522.80273      1381
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><div class="panel panel-default">
<div data-label="q7_ucsc" class="tutorial-question panel-body">
<div id="q7_ucsc-answer_container" class="shiny-html-output"></div>
<div id="q7_ucsc-message_container" class="shiny-html-output"></div>
<div id="q7_ucsc-action_button_container" class="shiny-html-output"></div>
<script>if (Tutorial.triggerMathJax) Tutorial.triggerMathJax()</script>
</div>
</div> <br> <strong>Later on in this tutorial we will learn how to annotate peaks with the gene that they most likely control. First we will calculate the fraction of overlap among two <code>GRanges</code> objects. </strong><br />
<br></p>
<hr />
<p><br></p>
</div>
</div>
<div id="section-overlap-analysis" class="section level2">
<h2>2.4 Overlap analysis</h2>
<div id="section-detect-overlap-with-genomicranges" class="section level3">
<h3>2.4.1 Detect overlap with GenomicRanges</h3>
<p>The GenomicRanges package has a family of functions to count and identify overlappig intervals in GRanges objects:<br />
<br> <strong>countOverlaps</strong>: <code>countOverlaps(query, subject)</code> returns a integer vector with the number of overlaps for each element in the <code>query</code><br />
<br> <strong>subsetByOverlaps</strong>: <code>subsetByOverlaps(query, subject)</code> extracts the elements in the <code>query</code> that overlap with at least one element in the subject.<br />
<br> <strong>findOverlaps</strong>: <code>findOverlaps(query, subject)</code> returns a <code>Hits</code> object containing the index pairings for the overlapping elements.</p>
<ul>
<li>The columns of indices can be accessed through <code>queryHits(overlap_object)</code> and <code>subjectHits(overlap_object)</code>.<br />
</li>
<li>If a peak in one of the inputs overlaps with mutiple peaks in the other, its index will appear multiple times in the output.</li>
</ul>
<p><br></p>
</div>
<div id="section-overlap-promoters-vs-h3k4me3-peaks" class="section level3">
<h3>2.4.2 Overlap promoters vs h3k4me3 peaks</h3>
<p>We have used the package <code>TxDb.Hsapiens.UCSC.hg38.knownGene</code> to retrieve the genomic coordinates of all genes on human chromosome 19. This object, <code>genes</code>, thus contains all the coordinates of the outermost UTR boundaries and the Entrez gene identifier of each gene.<br />
<br></p>
<pre class="r"><code># load the txdb package which holds transcript-based gene models of hg38 genome  
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb &lt;- TxDb.Hsapiens.UCSC.hg38.knownGene 

# extract gene coordinates  
genes &lt;- unique(genes(txdb, filter = list(tx_chrom = &quot;chr19&quot;)))</code></pre>
<p>We define the promoter region as 1kb upstream and 200bp downstream the TSS as follows:</p>
<pre class="r"><code># identify TSS (in this case 1 per gene even though we know that genes have multiple TSSs)
tss &lt;- GenomicRanges::resize(genes, width = 1, fix = &quot;start&quot;)

# consider 1kb upstream and 200bp downstream the TSS as promoter
promoters &lt;- GenomicRanges::resize(tss, width = 1000, fix = &quot;end&quot;)
promoters &lt;- GenomicRanges::resize(promoters, width = 1200, fix = &quot;start&quot;)

# show promoters object
head(promoters, n = 3)</code></pre>
<pre><code>## GRanges object with 3 ranges and 1 metadata column:
##             seqnames            ranges strand |     gene_id
##                &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##           1    chr19 58362551-58363750      - |           1
##   100049587    chr19 51646601-51647800      - |   100049587
##   100073347    chr19 56839867-56841066      + |   100073347
##   -------
##   seqinfo: 595 sequences (1 circular) from hg38 genome</code></pre>
<p><br> We want to know how many of the promoters overlap with a h3k4me3 peak and vise versa.<br />
<br></p>
<blockquote>
<p><strong>Exercise 8:</strong><br />
Use <code>findOverlaps()</code> to determine the overlap between (query) <code>mono_h3k4me3</code> and (subject) <code>promoters</code>.</p>
</blockquote>
<div class="tutorial-exercise-support" data-label="overlap_exercise7" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># prepare data for exercise 10
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb &lt;- TxDb.Hsapiens.UCSC.hg38.knownGene 

# extract gene coordinates  
genes &lt;- unique(genes(txdb, filter = list(tx_chrom = &quot;chr19&quot;)))

# identify TSS (in this case 1 per gene even though we know that genes have multiple TSSs)
tss &lt;- resize(genes, width = 1, fix = &quot;start&quot;)

# consider 1kb upstream and 200bp downstream the TSS as promoter
promoters &lt;- resize(tss, width = 1000, fix = &quot;end&quot;)
promoters &lt;- resize(promoters, width = 1200, fix = &quot;start&quot;)

# load the monocyte data
load(&quot;www/prepared_rds/blueprint_monocyte_chr19_granges.RData&quot;)</code></pre>
</div>
<div class="tutorial-exercise" data-label="q8_overlap" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Find overlap between mono_h3k4me3 peaks and promoters
overlap &lt;- ..(query = .., subject = ..)

# print the overlap output
show(..)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"overlap_exercise7","exercise.checker":"NULL"}</script>
</div>
<pre><code>## Hits object with 1256 hits and 0 metadata columns:
##          queryHits subjectHits
##          &lt;integer&gt;   &lt;integer&gt;
##      [1]         3        1609
##      [2]         5        1058
##      [3]         6         407
##      [4]         7         630
##      [5]         8         630
##      ...       ...         ...
##   [1252]      2833        1642
##   [1253]      2836        1341
##   [1254]      2836        1416
##   [1255]      2837        1341
##   [1256]      2837        1416
##   -------
##   queryLength: 2838 / subjectLength: 1729</code></pre>
<p><br> ‘overlap’ gives you the indices of <code>mono_h3k4me3</code> and <code>promoters</code> that overlap. If a peak or a promoter overlaps several times, each overlap will be reported on a new row.<br />
<br></p>
<blockquote>
<p><strong>Exercise 11:</strong><br />
How many of the promoters are part of the overlap? And how many of the h3k4me3 peaks?</p>
<ul>
<li>Use <code>queryHits()</code> and <code>subjectHits()</code> to extract the indices of overlapping peaks and promoters respectively.<br />
</li>
<li>Use <code>unique()</code> to minimize this output to unique peaks.<br />
</li>
<li>Use <code>length()</code> to count the number of unique intervals of the query and the subject.</li>
</ul>
</blockquote>
<div class="tutorial-exercise" data-label="q11a_overlap" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># What is the number of unique h3k4me3 peaks reported in &#39;overlap&#39;?
length(unique(..(..))) </code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"q10_overlap","exercise.checker":"NULL"}</script>
</div>
<pre><code>## [1] 1008</code></pre>
<div class="tutorial-exercise" data-label="q11b_overlap" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># What is the proportion of h3k4me3 peaks that overlap?
length(unique(..(..)))/length(..)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"q11_overlap","exercise.checker":"NULL"}</script>
</div>
<pre><code>## [1] 0.3551797</code></pre>
<div class="tutorial-exercise" data-label="q11c_overlap" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># What is the proportion of promoters that overlap?
length(unique(..(..)))/length(..)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"q11_overlap","exercise.checker":"NULL"}</script>
</div>
<pre><code>## [1] 0.6495084</code></pre>
<p><br> Let’s plot this overlap in a venn diagram with the <code>plot.pairwise.venn()</code> function of the <code>VennDiagram</code> package.</p>
<ul>
<li>Use the minimum of the two ‘unique counts’ as the number of ‘common peaks’ for our venn diagram.</li>
</ul>
<div class="tutorial-exercise" data-label="venndiagram" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Identify the minimum of the two numbers of unique intervals among h3k4me3 peaks and promoters
common_intervals &lt;- min(c(length(unique(queryHits(overlap))), length(unique(subjectHits(overlap)))))

# call a new plotting area
grid.newpage()

# Plot the overlap in a venn diagram
draw.pairwise.venn( 
   area1=length(mono_h3k4me3),
   area2=length(promoters), 
   cross.area=common_intervals, 
   category=c(&quot;h3k4me3&quot;, &quot;Promoters&quot;), 
   fill=c(&quot;red&quot;, &quot;gray&quot;), 
   cat.cex=1.2)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"q10_overlap","exercise.checker":"NULL"}</script>
</div>
<p><br> <strong>Is there a significant enrichment of h3k4me3 in promoters?</strong><br />
<br> To answer this question, we compare the fraction of promoters with a h3k4me3 peak with the chromosome 19-wide fraction of promoters.<br />
<br> If h3k4me3 is not enriched at promoters, we would expect that the fraction of promoters with a h3k4me3 peak is in the same range as the fraction of promoters on chromosome 19. (This is our <strong>null hypothesis of no enrichment</strong>).<br />
<br> To calculate the chromosome 19-wide fraction of promoters we identify the total number of bps covered by a promoter. As some promoters may overlap we first <em>reduce</em> <code>promoters</code> to non-overlapping intervals.<br />
<br></p>
<pre class="r"><code># total number of bps covered by promoters
total_bp_prom &lt;- sum(width(GenomicRanges::reduce(promoters)))
# chromosome 19-wide fraction of promoters 
(prom_fraction_chr19 &lt;- total_bp_prom/seqlengths(promoters)[&quot;chr19&quot;])</code></pre>
<pre><code>##      chr19 
## 0.03404707</code></pre>
<p><br> Promoters make up 3.4% of this chromosome but nearly 65% of all promoters overlap with h3k4me3 peaks. This looks like a strong enrichment. We can test whether the observed fraction is indeed larger than expected with a binomial test, in r we use the function <code>binom.test(x, n, p)</code> for this:</p>
<ul>
<li><code>x</code> = number of successes, number of promters with h3k4me3 peak<br />
</li>
<li><code>n</code> = total number of trails, total number of h3k4me3 peaks<br />
</li>
<li><code>p</code> = expected probability of success, in this case the fraction of promoters in chromosome 19</li>
</ul>
<p><br></p>
<blockquote>
<p><em>Background</em>: The binomial test is run when an experiment has two possible outcomes (i.e. success/failure) and you have an idea about what the probability of success is. Success in this case is overlap and our expectation is that 3.4% of the cases show overlap. The test calculates the probability of getting a desired outcome with a specific sample size <code>n</code>.</p>
</blockquote>
<blockquote>
<p><strong>Exercise 12</strong><br />
Use a binomial test to test for enrichment of h3k4me3 at promoters.</p>
<ul>
<li>Use the function <code>binom.test(x, n, p, alternative = "greater")</code> to call the test<br />
</li>
<li>We set <code>alternative = "greater"</code> because we test for <em>enrichment</em> and our alternative hypothesis is that the true probability is <em>larger</em> than the expected probability.</li>
</ul>
</blockquote>
<br>
<div class="tutorial-exercise" data-label="q12_binomtest" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># binomial test for enrichment of h3k4me3 peaks in promoters:  
binom.test(length(unique(..(overlap))), length(..), prom_fraction_chr19, alternative = &quot;greater&quot; )</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"fraction_promoters","exercise.checker":"NULL"}</script>
</div>
<pre><code>## 
##  Exact binomial test
## 
## data:  length(unique(subjectHits(overlap))) and length(mono_h3k4me3)
## number of successes = 1123, number of trials = 2838, p-value &lt;
## 2.2e-16
## alternative hypothesis: true probability of success is greater than 0.03404707
## 95 percent confidence interval:
##  0.3805112 1.0000000
## sample estimates:
## probability of success 
##              0.3957012</code></pre>
<p><br> The binomial test shows a p-value &lt; 2.2e-16 and a confidence interval that excludes the expected 0.034. We therefore reject the H0 of no enrichment.<br />
<br> In the genome browser we also observed H3K27ac often at promoters. Is this mark enriched in these regions?</p>
<blockquote>
<p><strong>Exercise 13</strong><br />
Use a binomial test to test for enrichment of H3K27ac at promoters.</p>
<ul>
<li>First, use findOverlaps to detect the overlap between intervals reported in mono_h3k37 and promoters</li>
<li>Use the function <code>binom.test(x, n, p, alternative = "greater")</code> to call the test<br />
</li>
<li>We set <code>alternative = "greater"</code> because we test for <em>enrichment</em> and our alternative hypothesis is that the true probability is <em>larger</em> than the expected probability.</li>
</ul>
</blockquote>
<div class="tutorial-exercise-support" data-label="overlap_exercise13" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># prepare data for exercise 13

# identify TSS (in this case 1 per gene even though we know that genes have multiple TSSs)
tss &lt;- resize(genes, width = 1, fix = &quot;start&quot;)
# consider 1kb upstream and 200bp downstream the TSS as promoter
promoters &lt;- resize(tss, width = 1000, fix = &quot;end&quot;)
promoters &lt;- resize(promoters, width = 1200, fix = &quot;start&quot;)

# load the monocyte data
load(&quot;www/prepared_rds/blueprint_monocyte_chr19_granges.RData&quot;)

# fraction of promoters on chr19
prom_fraction_chr19 &lt;- sum(width(GenomicRanges::reduce(promoters)))/seqlengths(promoters)[&quot;chr19&quot;]</code></pre>
</div>
<div class="tutorial-exercise" data-label="q13a_h3k27ac" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># find overlap between mono_h3k27ac and promoters
overlap2 &lt;- ...</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"overlap_exercise13","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="q13a_h3k27ac-solution" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>overlap2 &lt;- findOverlaps(query = mono_h3k27ac, subject = promoters)</code></pre>
</div>
<div class="tutorial-exercise" data-label="q13b_h3k27ac" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># binomial test for enrichment of H3K27ac peaks in promoters
# `prom_fraction_chr19` holds the fractio of promoters in chromosome 19
..(..(..(..(..))), ..(..), prom_fraction_chr19, alternative = &quot;greater&quot; )</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"q13a_h3k27ac","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="q13b_h3k27ac-solution" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># binomial test for enrichment of H3K27ac peaks in promoters:  
binom.test(length(unique(subjectHits(overlap2))), length(mono_h3k27ac), prom_fraction_chr19, alternative = &quot;greater&quot; )</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="q13b_h3k27ac-code-check" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>grade_code()</code></pre>
</div>
<p><br></p>
<blockquote>
<p><strong>Exercise 14:</strong><br />
Do you conclude that H3K27ac is enriched at promoters?</p>
</blockquote>
<div class="tutorial-exercise" data-label="q14_enrichment" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>question(&quot;Is H3K27ac enriched at promoters?&quot;,
         answer(&quot;*Yes*&quot;, correct = T),
         answer(&quot;*No*&quot;, message = &quot;Incorrect. Look at the p-value of the previous test.&quot;)
)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
</div>
<div id="section-overlap-with-multiple-genomic-features" class="section level3">
<h3>2.4.3 Overlap with multiple genomic features</h3>
<p><span style="color: red;">remove?</span><br />
To understand the function of a ChIPped mark or factor we often want to know to which genomic features it binds. <br></p>
<pre><code>## Loading required package: ChIPpeakAnno</code></pre>
<pre><code>## Warning: package &#39;ChIPpeakAnno&#39; was built under R version 3.6.1</code></pre>
<pre><code>## Loading required package: Biostrings</code></pre>
<pre><code>## Warning: package &#39;Biostrings&#39; was built under R version 3.6.1</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## Warning: package &#39;XVector&#39; was built under R version 3.6.1</code></pre>
<pre><code>## 
## Attaching package: &#39;XVector&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:purrr&#39;:
## 
##     compact</code></pre>
<pre><code>## 
## Attaching package: &#39;Biostrings&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     strsplit</code></pre>
<pre><code>## </code></pre>
<blockquote>
<p><strong>Exercise 13</strong><br />
Plot the percentage of features with the mark as barplot and as piechart</p>
<ul>
<li>Access the percent values in <code>h3k4me3_features</code> using the $ operator.<br />
</li>
<li>For plotting use <code>pie()</code> and <code>barplot()</code>.</li>
</ul>
</blockquote>
<div class="tutorial-exercise" data-label="q13a_anno" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># calculate the overlap with features
h3k4me3_features &lt;- assignChromosomeRegion(mono_h3k4me3, TxDb=txdb, precedence = c(&quot;Promoters&quot;,&quot;immediateDownstream&quot;, &quot;fiveUTRs&quot;, &quot;threeUTRs&quot;, &quot;Exons&quot;, &quot;Introns&quot; ), nucleotideLevel=FALSE)
# show the results
h3k4me3_features

# make pie-chart</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<pre><code>## $percentage
## subjectHits
##           Promoters immediateDownstream            fiveUTRs 
##           15.856237            7.822410            2.290345 
##           threeUTRs               Exons             Introns 
##            2.396054            9.654686           33.474278 
##   Intergenic.Region 
##           28.505990 
## 
## $jaccard
##           Promoters immediateDownstream            fiveUTRs 
##          0.08682230          0.05611144          0.05285222 
##           threeUTRs               Exons             Introns 
##          0.04532164          0.05347753          0.12508779 
##   Intergenic.Region 
##          0.27787406</code></pre>
<p><img src="FunctGenWeek2_v2_files/figure-html/q13a_anno-solution-1.png" width="624" /></p>
<div class="tutorial-exercise" data-label="q13b_anno" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># make barplot
bp &lt;- ..(..$.., ylab=&quot;%&quot;)
# add percentages to the bars with text()  
text(bp, h3k4me3_features$percentage, signif(h3k4me3_features$percentage, 4), pos=1, main = &quot;Mono h3k4me3 in features&quot;)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"anno_h3k4me3_prep","exercise.checker":"NULL"}</script>
</div>
<p><img src="FunctGenWeek2_v2_files/figure-html/q13anno-solution-1.png" width="624" /></p>
</div>
<div id="section-overlap-with-grangeslist" class="section level3">
<h3>2.4.5 Overlap with GRangesList</h3>
<p><span style="color: red;">remove?</span><br />
So far we looked at h3k4me3 marking. How do the other marks overlap with genomic features? Instead of rerunning the above code for each mark individually, we combine the individual peak objects into a list of <code>GRanges</code> objects with the function <code>GRangesList()</code>.<br />
<br><br />
We have created a <code>GRangesList</code> <code>mono_list</code> by:</p>
<pre class="r"><code># make a list of your GRanges objects
mono_list  &lt;- GRangesList(mono_h3k4me1, mono_h3k4me3, mono_h3k9me3, mono_h3k27ac, mono_h3k27me3, mono_h3k36me3)
# add names to each element in the list
names(mono_list) &lt;- c(&quot;h3k4me1&quot;, &quot;h3k4me3&quot;, &quot;h3k9me3&quot;, &quot;h3k27ac&quot;, &quot;h3k27me3&quot;, &quot;h3k36me3&quot;)</code></pre>
<pre class="r"><code>load(paste0(datadir, &quot;www/prepared_rds/txdb_annotation_chr19_genes_tx_pr_tss_granges.RData&quot;))

# print the length of your list  
length(mono_list)</code></pre>
<pre><code>## [1] 6</code></pre>
<p><br> The <code>length</code> function now applies to the list and not to the individual items any more.<br />
<br> We can access elements in the list using <code>[[]]</code> or <code>$</code>-sign.</p>
<ul>
<li>Use <code>[[]]</code> and the <em>index</em>: eg. <code>mono_list[[1]]</code> will extract the first item in the list.</li>
<li>Use <code>[[]]</code> and the name: eg <code>mono_list[["h3k36me3"]]</code> will extract the item with this name.<br />
</li>
<li>Use <code>$item_name</code> to extract the item with that name.</li>
</ul>
<p>To calculate the length of individual items in the list use <code>lapply()</code>, which applies a function to each item in the list.<br />
<br> With the fuction <code>lapply(list_object, function)</code> we can apply or <em>loop</em> a function to each element in the list and returns the results as a list.<br />
<br> Here we apply <code>length()</code> to each item in <code>mono_list</code></p>
<pre class="r"><code># print the length of each element at once with apply
lapply(mono_list, length) </code></pre>
<pre><code>## $h3k4me1
## [1] 3282
## 
## $h3k4me3
## [1] 2838
## 
## $h3k9me3
## [1] 1893
## 
## $h3k27ac
## [1] 3793
## 
## $h3k27me3
## [1] 3059
## 
## $h3k36me3
## [1] 2084</code></pre>
<p><br> <code>unlist()</code> flattens the list to one object. Depending on the elements in the list this will return a <code>vector</code>, <code>matrix</code> or, if all elements have the same data structure and column names, one large version of that particular data structure.<br />
<br> E.g. <code>unlist(mono_list)</code> will give you one large <code>GRanges</code> object that is often not usefull because peaks of all marks are mixed up. <code>unlist(lapply(mono_list, length))</code>, on the other hand, returns a usefull vector of peak numbers per ChIP.<br />
<br></p>
<blockquote>
<p><strong>Exercise 14:</strong><br />
Make a barplot of peak numbers per ChIP:**</p>
<ul>
<li>Use the <code>mono_list</code> object<br />
</li>
<li>Obtain <code>counts_vector</code> with <code>lapply()</code>, and <code>unlist()</code><br />
</li>
<li>Use the <code>length</code> within <code>lapply()</code></li>
</ul>
</blockquote>
<div class="tutorial-exercise" data-label="q14_barplot" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># print the length of each element at once with apply
counts_list &lt;- .. ( .. , .. )

# unlist the resulting output
counts_vector &lt;- ..(..)

# make barplot of the number of peaks, store this in &#39;bp&#39;:  
bp &lt;- ..(.., ylab = &quot;Number of peaks&quot;, main = &quot;Monocytes ChIP-seq; chr19&quot;)

# add actual values as text lables to the plot
text(bp, counts_vector, labels=names(counts_vector), pos=1)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<p><img src="FunctGenWeek2_v2_files/figure-html/q14_barplot-solution-1.png" width="624" /></p>
<p><br> Instead of the length, let’s look at the distribution of these peaks among features, as in section 2.4.3.<br />
<br></p>
<blockquote>
<p><strong>Exercise 15</strong></p>
<p>For each item in <code>mono_list</code> determine the overlap with genomic features. Use the <code>assignChromosomeRegion(object,TxDb=txdb, nucleotideLevel=FALSE)</code> to assign peaks to features.</p>
</blockquote>
<pre><code>## $h3k4me1
## $h3k4me1$percentage
## subjectHits
##               Exons             Introns            fiveUTRs 
##            37.99512            64.04631            15.66118 
##           threeUTRs           Promoters immediateDownstream 
##            11.63924            18.46435            14.68617 
##   Intergenic.Region 
##            27.11761 
## 
## $h3k4me1$jaccard
##               Exons             Introns            fiveUTRs 
##          0.08055035          0.16283213          0.07648810 
##           threeUTRs           Promoters immediateDownstream 
##          0.07698509          0.11468584          0.09312210 
##   Intergenic.Region 
##          0.33093717 
## 
## 
## $h3k4me3
## $h3k4me3$percentage
## subjectHits
##               Exons             Introns            fiveUTRs 
##            31.53629            62.75546            12.43834 
##           threeUTRs           Promoters immediateDownstream 
##             7.64623            15.85624            10.07752 
##   Intergenic.Region 
##            28.50599 
## 
## $h3k4me3$jaccard
##               Exons             Introns            fiveUTRs 
##          0.05347753          0.12508779          0.05285222 
##           threeUTRs           Promoters immediateDownstream 
##          0.04532164          0.08682230          0.05611144 
##   Intergenic.Region 
##          0.27787406 
## 
## 
## $h3k9me3
## $h3k9me3$percentage
## subjectHits
##               Exons             Introns            fiveUTRs 
##            33.59746            54.67512            15.79503 
##           threeUTRs           Promoters immediateDownstream 
##            13.89329            16.58743            16.69308 
##   Intergenic.Region 
##            37.82356 
## 
## $h3k9me3$jaccard
##               Exons             Introns            fiveUTRs 
##          0.04180360          0.07841503          0.05498345 
##           threeUTRs           Promoters immediateDownstream 
##          0.07170120          0.07358800          0.07933718 
##   Intergenic.Region 
##          0.38971161 
## 
## 
## $h3k27ac
## $h3k27ac$percentage
## subjectHits
##               Exons             Introns            fiveUTRs 
##           22.462431           63.828104            7.487477 
##           threeUTRs           Promoters immediateDownstream 
##            4.877406           11.468495            7.592934 
##   Intergenic.Region 
##           27.339837 
## 
## $h3k27ac$jaccard
##               Exons             Introns            fiveUTRs 
##          0.04691888          0.15459770          0.03645700 
##           threeUTRs           Promoters immediateDownstream 
##          0.03188006          0.06926752          0.04707421 
##   Intergenic.Region 
##          0.25216619 
## 
## 
## $h3k27me3
## $h3k27me3$percentage
## subjectHits
##               Exons             Introns            fiveUTRs 
##           28.146453           56.652501            9.774436 
##           threeUTRs           Promoters immediateDownstream 
##            8.695652           12.389670           10.526316 
##   Intergenic.Region 
##           35.436417 
## 
## $h3k27me3$jaccard
##               Exons             Introns            fiveUTRs 
##          0.05198961          0.12309113          0.04308357 
##           threeUTRs           Promoters immediateDownstream 
##          0.05431897          0.06942663          0.06166220 
##   Intergenic.Region 
##          0.36268575 
## 
## 
## $h3k36me3
## $h3k36me3$percentage
## subjectHits
##               Exons             Introns            fiveUTRs 
##            45.72937            69.28983            21.54511 
##           threeUTRs           Promoters immediateDownstream 
##            18.61804            20.96929            21.11324 
##   Intergenic.Region 
##            24.42418 
## 
## $h3k36me3$jaccard
##               Exons             Introns            fiveUTRs 
##          0.07177286          0.12766334          0.08331787 
##           threeUTRs           Promoters immediateDownstream 
##          0.10703448          0.10323648          0.11210191 
##   Intergenic.Region 
##          0.28888081</code></pre>
</div>
</div>
<div id="section-annotation-of-chip-seq-peaks" class="section level2">
<h2>2.5 Annotation of ChIP-seq peaks</h2>
<div id="section-distance-to-tss" class="section level3">
<h3>2.5.1 Distance to TSS</h3>
<p>In this part of the analysis we will assign peaks to their closest TSS, assuming that it likely is involved in regulating the expression regulates that gene.<br />
<br> The <code>GenomicRanges</code> package has the function <code>distanceToNearest(x, subject)</code> to help you identify the distance between intervals in <code>x</code> and the nearest neighbor in <code>subject</code>. In this case, the intervals are the peaks and the subject are the TSSs. The function outputs a <code>Hits</code> objects with distances.<br />
<br></p>
<blockquote>
<p><strong>Exercise 15</strong><br />
Use <code>distanceToNearest()</code> to get the distance between h3k4me3 peaks and the nearest TSS.</p>
</blockquote>
<div class="tutorial-exercise" data-label="q15_distance_tss" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># calculate the distances from peaks to tss
h3k4me3_to_tss &lt;- ..(.., ..)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"overlap_exercise10","exercise.checker":"NULL"}</script>
</div>
<p><br> Let’s look at the resulting <code>Hits</code> object.<br />
<br></p>
<pre class="r"><code># show object h3k4me3_to_tss
show(h3k4me3_to_tss)</code></pre>
<pre><code>## Hits object with 2838 hits and 1 metadata column:
##          queryHits subjectHits |  distance
##          &lt;integer&gt;   &lt;integer&gt; | &lt;integer&gt;
##      [1]         1         924 |     36487
##      [2]         2        1609 |     23749
##      [3]         3        1609 |         0
##      [4]         4        1609 |      3603
##      [5]         5        1058 |         0
##      ...       ...         ... .       ...
##   [2834]      2834          25 |      2812
##   [2835]      2835          25 |      3247
##   [2836]      2836        1341 |         0
##   [2837]      2837        1416 |        53
##   [2838]      2838        1416 |      1545
##   -------
##   queryLength: 2838 / subjectLength: 1729</code></pre>
<pre class="r"><code># summary of distances
summary(mcols(h3k4me3_to_tss)[,1])</code></pre>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##       0       0    3422    9599   11567  447135</code></pre>
<pre class="r"><code># plot the distribution of distances
boxplot(mcols(h3k4me3_to_tss)[,1])</code></pre>
<p><img src="FunctGenWeek2_v2_files/figure-html/look%20at%20hits-1.png" width="624" /> <br></p>
</div>
<div id="section-find-associating-gene" class="section level3">
<h3>2.5.2 Find associating gene</h3>
<p>We can use this <code>Hits</code> object to subset peaks to those that are less than 10 Kb away from the TSS. We can than use the function <code>subjectHits()</code> to get the genes associated to each peak.<br />
<br></p>
<blockquote>
<p><strong>Exercise 16</strong><br />
Subset the h3k4me3_to_tss for distances =&lt; 10kb and select the associated genes from <code>genes</code> object.</p>
</blockquote>
<div class="tutorial-exercise-support" data-label="prepare_closegene_q16" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>load(&quot;www/prepared_rds/txdb_annotation_chr19_genes_tx_pr_tss_granges.RData&quot;)  
h3k4me3_to_tss &lt;- distanceToNearest(mono_h3k4me3, tss)</code></pre>
</div>
<div class="tutorial-exercise" data-label="q16a_closegenes" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># save the distances in a vector
distances &lt;- mcols(h3k4me3_to_tss)[,1]

# subset h3k4me3_to_tss by distance
h3k4me3_to_tss_close &lt;- ..[..,..]</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"prepare_closegene_q16","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise" data-label="16b_closegenes" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># obtain the genes
mono_h3k4me3_genes &lt;- unique(genes[..(..)])

# look at the output
show(mono_h3k4me3_genes)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"q16a_closegenes","exercise.checker":"NULL"}</script>
</div>
<pre><code>## GRanges object with 1144 ranges and 1 metadata column:
##             seqnames            ranges strand |     gene_id
##                &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##        8612    chr19     281040-291403      - |        8612
##       54531    chr19     301444-344815      - |       54531
##      126567    chr19     405445-409147      - |      126567
##       25759    chr19     416583-460996      - |       25759
##      284451    chr19     463346-474983      - |      284451
##         ...      ...               ...    ... .         ...
##       27243    chr19 58551566-58555105      - |       27243
##        9040    chr19 58555712-58558954      - |        9040
##   100131691    chr19 58559129-58574797      + |   100131691
##       65996    chr19 58573503-58605223      + |       65996
##        7593    chr19 58561932-58573575      - |        7593
##   -------
##   seqinfo: 1 sequence from hg38 genome</code></pre>
<p><br> Besides the distance, we are also (very!) interested in the distribution of the ChIP-seq signal around the TSS. We will look at that next week.</p>
</div>
</div>
<div id="section-background" class="section level2">
<h2>background</h2>
<div id="section-background-bioconductor" class="section level3">
<h3>2.2.1 Background Bioconductor</h3>
<p><a href="https://www.bioconductor.org/">Bioconductor</a> is a special repository of packages for the analysis of high throughput genomic data in r. Packages are often heavily dependend on each other and regular releases and a special installer ensure that you install package version that can work together. Usually you would use the <code>biocManager::install()</code> command to install these packages. But in this tutorial, all required packages are already installed.<br />
<br> Bioconductor packages that we will be using include:</p>
<table>
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<thead>
<tr class="header">
<th>Package</th>
<th>Purpose</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">GenomicRanges</a></td>
<td>Provides the <code>GRanges</code> data structures to store and handle genomic intervals.</td>
</tr>
<tr class="even">
<td><a href="https://bioconductor.org/packages/release/bioc/html/plyranges.html">plyranges</a></td>
<td><code>dplyr</code>-like interface for interacting and manipulating <code>Ranges</code> data structures.</td>
</tr>
<tr class="odd">
<td><a href="https://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html">GenomicFeatures</a></td>
<td>Functions to retrieve and manage genomic features from database packages.</td>
</tr>
<tr class="even">
<td><a href="https://www.bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a></td>
<td>Provides functions to import and work with annotation files with various formats (GFF, BED, bedGraph, BED15, WIG, BigWig and 2bit).</td>
</tr>
<tr class="odd">
<td><code>TxDb</code> packages</td>
<td>Provide an R representation of gene models, each genome has a separate package.</td>
</tr>
<tr class="even">
<td><code>OrgDb</code> packages</td>
<td>Contain mappings between a unique gene identifier and other kinds of identifiers for a certain genome (eg from Entrez gene identifier to Gene Symbol)</td>
</tr>
</tbody>
</table>
<p><br> Bioconductor packages can use <strong>data structures</strong> different from <code>vectors</code>,<code>matrices</code> or <code>data.frames</code>. The <code>GRanges</code> data structure provided by the <code>GenomicRanges</code> package is an example and is used by many. We will come to those shortly.<br />
<br> These “Bioconductor-specific-objects”, like <code>GRanges</code>, generally have their own set of functions or methods that you can explore using <code>methods(class = "...")</code>. To find out the <code>class</code>, use <code>class(object_name)</code>.<br />
<br> Here’s a list of basic functions to explore objects and/or call the help on packages or functions:</p>
<ul>
<li>Use <code>class()</code> to found out what kind of data structure you are dealing with<br />
</li>
<li>Use <code>show()</code> or <code>print()</code> to have r print a summary of your data<br />
</li>
<li>use <code>methods(class = "...")</code> to get a list of (default) methods or functions you can use to extract and manipulate the data, also called <em>assessor</em> functions or <em>getter</em> and <em>setter</em> functions<br />
</li>
<li>use <code>help(package = "[packageName]")</code> to display the help page</li>
</ul>
<p><br></p>
<hr />
<p><br></p>

<script type="application/shiny-prerendered" data-context="server-start">
# packages
library(learnr)  # 0.10.1 -> conda-forge / 0.9.2.1 -> r
library(tidyverse)  # 1.2.1 -> r / 1.3.0 -> conda-forge
library(ggplot2)  # 3.1.1 -> r
#installed required libraries via .yaml file environment_v3_withversions.yml
#see Home/GitHub/learnr_20201020_installing_learnr_conda_environment.txt
library(IRanges)
library(tidyverse)
library(GenomicRanges)
library(plyranges)
library(rtracklayer)
library(VennDiagram)
#library(UpSetR)
library(GenomicFeatures)
#library(ChIPpeakAnno)
#library(genomation)
source("functions.r")

# configuration
knitr::opts_chunk$set(echo = FALSE)
# gradethis::gradethis_setup()

# functions
# checker <- function(label, user_code, check_code, envir_result, evaluate_result, ...) {
#  list(message = check_code, correct = TRUE, location = "append")
# }
#tutorial_options(
#  exercise.timelimit = 10,  # max code execution time
  #exercise.checker = checker)

# # what to do when the browser window is closed
# session$onSessionEnded(function() {
# 
#   # TODO save progress
# 
#   stopApp()
#   })

</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::register_http_handlers(session, metadata = list(id = "prototype", version = 0.1))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
session$onSessionEnded(function() {
        learnr:::session_stop_event(session)
      })
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q1_head_h3k4me3-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q1_head_h3k4me3-code-editor`)), session)
output$`tutorial-exercise-q1_head_h3k4me3-output` <- renderUI({
  `tutorial-exercise-q1_head_h3k4me3-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q2a_explore-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q2a_explore-code-editor`)), session)
output$`tutorial-exercise-q2a_explore-output` <- renderUI({
  `tutorial-exercise-q2a_explore-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q2b_explore-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q2b_explore-code-editor`)), session)
output$`tutorial-exercise-q2b_explore-output` <- renderUI({
  `tutorial-exercise-q2b_explore-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q2c_explore-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q2c_explore-code-editor`)), session)
output$`tutorial-exercise-q2c_explore-output` <- renderUI({
  `tutorial-exercise-q2c_explore-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q3a_exploreplot-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q3a_exploreplot-code-editor`)), session)
output$`tutorial-exercise-q3a_exploreplot-output` <- renderUI({
  `tutorial-exercise-q3a_exploreplot-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q3b_exploreplot-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q3b_exploreplot-code-editor`)), session)
output$`tutorial-exercise-q3b_exploreplot-output` <- renderUI({
  `tutorial-exercise-q3b_exploreplot-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q4_filter-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q4_filter-code-editor`)), session)
output$`tutorial-exercise-q4_filter-output` <- renderUI({
  `tutorial-exercise-q4_filter-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q5_filter-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q5_filter-code-editor`)), session)
output$`tutorial-exercise-q5_filter-output` <- renderUI({
  `tutorial-exercise-q5_filter-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q6_maxscore-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q6_maxscore-code-editor`)), session)
output$`tutorial-exercise-q6_maxscore-output` <- renderUI({
  `tutorial-exercise-q6_maxscore-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::question_prerendered_chunk(structure(list(type = "learnr_radio", label = "q7_ucsc", question = structure("The promoters of which gene(s) are covered by the h3k4me3 peak with the highest score?", html = TRUE, class = c("html", "character")), answers = list(structure(list(id = "lnr_ans_cc1d5eb",     option = "AKAP8", value = "AKAP8", label = structure("AKAP8", html = TRUE, class = c("html",     "character")), correct = TRUE, message = structure("While this peak spans parts of both genes, it covers the promoter of AKAP8 but the 3&#39;UTR of AKAP8L.", html = TRUE, class = c("html",     "character"))), class = c("tutorial_question_answer", "tutorial_quiz_answer")), structure(list(id = "lnr_ans_c13060f", option = "AKAP8L",     value = "AKAP8L", label = structure("AKAP8L", html = TRUE, class = c("html",     "character")), correct = FALSE, message = structure("Incorrect. Be aware of the directionaility of this gene by looking at the arrowsheads in the introns.", html = TRUE, class = c("html",     "character"))), class = c("tutorial_question_answer", "tutorial_quiz_answer")), structure(list(id = "lnr_ans_832d0c6", option = "Both", value = "Both",     label = structure("Both", html = TRUE, class = c("html",     "character")), correct = FALSE, message = structure("Only one of the genes listed here. Look at the direction of transcription indicated by the arrowheads in the introns of the genes.", html = TRUE, class = c("html",     "character"))), class = c("tutorial_question_answer", "tutorial_quiz_answer"))), button_labels = list(submit = structure("Submit Answer", html = TRUE, class = c("html", "character")), try_again = structure("Try Again", html = TRUE, class = c("html", "character"))), messages = list(correct = structure("Correct!", html = TRUE, class = c("html", "character")), try_again = structure("Incorrect", html = TRUE, class = c("html", "character")), incorrect = structure("Incorrect", html = TRUE, class = c("html", "character")), message = NULL, post_message = NULL), ids = list(    answer = "q7_ucsc-answer", question = "q7_ucsc"), loading = structure("<strong>Loading:<\u002fstrong> \nThe promoters of which gene(s) are covered by the h3k4me3 peak with the highest score?\n<br/><br/><br/>", html = TRUE, class = c("html", "character")), random_answer_order = FALSE, allow_retry = FALSE,     seed = 485650102.273852, options = list()), class = c("learnr_radio", "tutorial_question")))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q8_overlap-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q8_overlap-code-editor`)), session)
output$`tutorial-exercise-q8_overlap-output` <- renderUI({
  `tutorial-exercise-q8_overlap-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q11a_overlap-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q11a_overlap-code-editor`)), session)
output$`tutorial-exercise-q11a_overlap-output` <- renderUI({
  `tutorial-exercise-q11a_overlap-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q11b_overlap-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q11b_overlap-code-editor`)), session)
output$`tutorial-exercise-q11b_overlap-output` <- renderUI({
  `tutorial-exercise-q11b_overlap-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q11c_overlap-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q11c_overlap-code-editor`)), session)
output$`tutorial-exercise-q11c_overlap-output` <- renderUI({
  `tutorial-exercise-q11c_overlap-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-venndiagram-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-venndiagram-code-editor`)), session)
output$`tutorial-exercise-venndiagram-output` <- renderUI({
  `tutorial-exercise-venndiagram-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q12_binomtest-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q12_binomtest-code-editor`)), session)
output$`tutorial-exercise-q12_binomtest-output` <- renderUI({
  `tutorial-exercise-q12_binomtest-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q13a_h3k27ac-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q13a_h3k27ac-code-editor`)), session)
output$`tutorial-exercise-q13a_h3k27ac-output` <- renderUI({
  `tutorial-exercise-q13a_h3k27ac-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q13b_h3k27ac-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q13b_h3k27ac-code-editor`)), session)
output$`tutorial-exercise-q13b_h3k27ac-output` <- renderUI({
  `tutorial-exercise-q13b_h3k27ac-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q14_enrichment-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q14_enrichment-code-editor`)), session)
output$`tutorial-exercise-q14_enrichment-output` <- renderUI({
  `tutorial-exercise-q14_enrichment-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q13a_anno-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q13a_anno-code-editor`)), session)
output$`tutorial-exercise-q13a_anno-output` <- renderUI({
  `tutorial-exercise-q13a_anno-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q13b_anno-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q13b_anno-code-editor`)), session)
output$`tutorial-exercise-q13b_anno-output` <- renderUI({
  `tutorial-exercise-q13b_anno-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q14_barplot-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q14_barplot-code-editor`)), session)
output$`tutorial-exercise-q14_barplot-output` <- renderUI({
  `tutorial-exercise-q14_barplot-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q15_distance_tss-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q15_distance_tss-code-editor`)), session)
output$`tutorial-exercise-q15_distance_tss-output` <- renderUI({
  `tutorial-exercise-q15_distance_tss-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-q16a_closegenes-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-q16a_closegenes-code-editor`)), session)
output$`tutorial-exercise-q16a_closegenes-output` <- renderUI({
  `tutorial-exercise-q16a_closegenes-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-16b_closegenes-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-16b_closegenes-code-editor`)), session)
output$`tutorial-exercise-16b_closegenes-output` <- renderUI({
  `tutorial-exercise-16b_closegenes-result`()
})
</script>
 <!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootstrap"]},{"type":"character","attributes":{},"value":["3.3.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/bootstrap"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["viewport"]}},"value":[{"type":"character","attributes":{},"value":["width=device-width, initial-scale=1"]}]},{"type":"character","attributes":{},"value":["js/bootstrap.min.js","shim/html5shiv.min.js","shim/respond.min.js"]},{"type":"character","attributes":{},"value":["css/cerulean.min.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["pagedtable"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pagedtable-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["js/pagedtable.js"]},{"type":"character","attributes":{},"value":["css/pagedtable.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["textmate.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-format"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmarkdown/templates/tutorial/resources"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-format.js"]},{"type":"character","attributes":{},"value":["tutorial-format.css","rstudio-theme.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["navigation"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/navigation-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tabsets.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["default.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["font-awesome"]},{"type":"character","attributes":{},"value":["5.1.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/fontawesome"]}]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["css/all.css","css/v4-shims.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.12"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootbox"]},{"type":"character","attributes":{},"value":["4.4.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/bootbox"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["bootbox.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["idb-keyvalue"]},{"type":"character","attributes":{},"value":["3.2.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/idb-keyval"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["idb-keyval-iife-compat.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141]}},"value":[{"type":"character","attributes":{},"value":["AnnotationDbi","AnnotationFilter","BSgenome","Biobase","BiocFileCache","BiocGenerics","BiocManager","BiocParallel","Biostrings","ChIPpeakAnno","DBI","DelayedArray","GO.db","GenomeInfoDb","GenomeInfoDbData","GenomicAlignments","GenomicFeatures","GenomicRanges","IRanges","MASS","Matrix","ProtGenerics","R6","RBGL","RCurl","RSQLite","Rcpp","Rsamtools","S4Vectors","SummarizedExperiment","TxDb.Hsapiens.UCSC.hg19.knownGene","TxDb.Hsapiens.UCSC.hg38.knownGene","VennDiagram","XML","XVector","ade4","askpass","assertthat","backports","base","biomaRt","bit","bit64","bitops","blob","broom","cellranger","checkmate","cli","colorspace","compiler","crayon","curl","datasets","dbplyr","digest","dplyr","ellipsis","ensembldb","evaluate","forcats","formatR","futile.logger","futile.options","generics","ggplot2","glue","grDevices","graph","graphics","grid","gtable","haven","hms","htmltools","htmlwidgets","httpuv","httr","idr","jsonlite","knitr","lambda.r","later","lattice","lazyeval","learnr","lifecycle","limma","lubridate","magrittr","markdown","matrixStats","memoise","methods","mime","modelr","multtest","munsell","nlme","openssl","parallel","pillar","pkgconfig","plyr","plyranges","prettyunits","progress","promises","purrr","rappdirs","readr","readxl","regioneR","rlang","rmarkdown","rprojroot","rstudioapi","rtracklayer","rvest","scales","seqinr","shiny","splines","stats","stats4","stringi","stringr","survival","tibble","tidyr","tidyselect","tidyverse","tools","utils","vctrs","withr","xfun","xml2","xtable","yaml","zlibbioc"]},{"type":"character","attributes":{},"value":["1.48.0","1.10.0","1.54.0","2.46.0","1.10.0","0.32.0","1.30.4","1.20.0","2.54.0","3.20.0","1.0.0","0.12.0","3.10.0","1.22.0","1.2.2","1.22.0","1.38.0","1.38.0","2.20.0","7.3-51.3","1.2-17","1.18.0","2.4.0","1.62.1","1.95-4.12","2.1.1","1.0.1","2.2.0","0.24.0","1.16.0","3.2.2","3.10.0","1.6.20","3.98-1.19","0.26.0","1.7-13","1.0","0.2.1","1.1.4","3.6.0","2.42.0","1.1-14","0.9-7","1.0-6","1.1.1","0.5.2","1.1.0","1.9.1","1.1.0","1.4-1","3.6.0","1.3.4","3.3","3.6.0","1.4.0","0.6.18","0.8.5","0.3.1","2.10.0","0.13","0.4.0","1.6","1.4.3","1.0.1","0.0.2","3.1.1","1.3.1","3.6.0","1.64.0","3.6.0","3.6.0","0.3.0","2.1.0","0.4.2","0.3.6","1.3","1.5.1","1.4.0","1.2","1.6","1.22","1.2.3","0.8.0","0.20-38","0.2.2","0.10.1","0.2.0","3.42.0","1.7.4","1.5","0.9","0.54.0","1.1.0","3.6.0","0.6","0.1.4","2.42.0","0.5.0","3.1-139","1.3","3.6.0","1.3.1","2.0.2","1.8.4","1.6.0","1.0.2","1.2.0","1.0.1","0.3.2","0.3.1","1.3.1","1.3.1","1.18.0","0.4.8","1.12","1.3-2","0.10","1.46.0","0.3.3","1.0.0","4.2-4","1.3.2","3.6.0","3.6.0","3.6.0","1.4.3","1.4.0","2.44-1.1","2.1.1","1.1.2","1.1.0","1.2.1","3.6.0","3.6.0","0.3.5","2.1.2","0.6","1.2.0","1.8-4","2.2.0","1.32.0"]}]}]}
</script>
<!--/html_preserve-->
</div>
</div>

</div> <!-- topics -->

<div class="topicsContainer">
<div class="topicsPositioner">
<div class="band">
<div class="bandContent topicsListContainer">

<!-- begin doc-metadata -->
<div id="doc-metadata">
<h2 class="title toc-ignore" style="display:none;">Week 2: Exploratory &amp; Overlap analysis</h2>
</div>
<!-- end doc-metadata -->

</div> <!-- bandContent.topicsListContainer -->
</div> <!-- band -->
</div> <!-- topicsPositioner -->
</div> <!-- topicsContainer -->


</div> <!-- bandContent page -->
</div> <!-- pageContent band -->




<script>
// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>


</body>

</html>
